

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>tidymut.core.dataset &mdash; tidymut 0.1.0-dev documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=ee2d09ae"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            tidymut
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../tidymut.cleaners.html">tidymut.cleaners package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tidymut.core.html">tidymut.core package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tidymut.utils.html">tidymut.utils package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">tidymut</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">tidymut.core.dataset</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for tidymut.core.dataset</h1><div class="highlight"><pre>
<span></span><span class="c1"># tidymut/core/dataset.py</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">__future__</span><span class="w"> </span><span class="kn">import</span> <span class="n">annotations</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pickle</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">json</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tqdm</span><span class="w"> </span><span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">cast</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">TYPE_CHECKING</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">.mutation</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">MutationSet</span><span class="p">,</span>
    <span class="n">AminoAcidMutationSet</span><span class="p">,</span>
    <span class="n">CodonMutationSet</span><span class="p">,</span>
    <span class="n">AminoAcidMutation</span><span class="p">,</span>
    <span class="n">CodonMutation</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">.sequence</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">DNASequence</span><span class="p">,</span>
    <span class="n">ProteinSequence</span><span class="p">,</span>
    <span class="n">RNASequence</span><span class="p">,</span>
    <span class="n">load_sequences_from_fasta</span><span class="p">,</span>
<span class="p">)</span>

<span class="k">if</span> <span class="n">TYPE_CHECKING</span><span class="p">:</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Literal</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">Type</span><span class="p">,</span> <span class="n">Union</span>

    <span class="kn">from</span><span class="w"> </span><span class="nn">.mutation</span><span class="w"> </span><span class="kn">import</span> <span class="n">BaseMutation</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">.sequence</span><span class="w"> </span><span class="kn">import</span> <span class="n">BaseSequence</span>

<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;MutationDataset&quot;</span><span class="p">]</span>


<span class="k">def</span><span class="w"> </span><span class="fm">__dir__</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
    <span class="k">return</span> <span class="n">__all__</span>


<span class="n">SEQUENCE_TYPE_MAP</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;ProteinSequence&quot;</span><span class="p">:</span> <span class="n">ProteinSequence</span><span class="p">,</span>
    <span class="s2">&quot;DNASequence&quot;</span><span class="p">:</span> <span class="n">DNASequence</span><span class="p">,</span>
    <span class="s2">&quot;RNASequence&quot;</span><span class="p">:</span> <span class="n">RNASequence</span><span class="p">,</span>
    <span class="s2">&quot;DNA&quot;</span><span class="p">:</span> <span class="n">DNASequence</span><span class="p">,</span>
    <span class="s2">&quot;RNA&quot;</span><span class="p">:</span> <span class="n">RNASequence</span><span class="p">,</span>
    <span class="s2">&quot;Protein&quot;</span><span class="p">:</span> <span class="n">ProteinSequence</span><span class="p">,</span>
<span class="p">}</span>


<div class="viewcode-block" id="MutationDataset">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">MutationDataset</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Dataset container for cleaned mutation data with multiple reference sequences.</span>

<span class="sd">    All mutation sets must be linked to a reference sequence when added to the dataset.</span>
<span class="sd">    This ensures data integrity and enables proper validation and analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize a new MutationDataset.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            name: Optional name for the dataset</span>

<span class="sd">        Note:</span>
<span class="sd">            All mutation sets added to this dataset must be linked to a reference sequence.</span>
<span class="sd">            Use add_reference_sequence() first, then add_mutation_set() with reference_id.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BaseSequence</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">{}</span>
        <span class="p">)</span>  <span class="c1"># sequence_id -&gt; sequence</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">MutationSet</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">{}</span>
        <span class="p">)</span>  <span class="c1"># mutation_set_index -&gt; sequence_id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># mutation_set_index -&gt; label</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_df</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Iterate over mutation sets and their reference sequence IDs.</span>

<span class="sd">        Yields:</span>
<span class="sd">            Tuple[MutationSet, str]: (mutation_set, reference_id) pairs</span>

<span class="sd">        Example:</span>
<span class="sd">            for mutation_set, ref_id in dataset:</span>
<span class="sd">                print(f&quot;Processing {len(mutation_set)} mutations for {ref_id}&quot;)</span>
<span class="sd">                ref_seq = dataset.get_reference_sequence(ref_id)</span>
<span class="sd">                # ... analysis code</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mutation_set</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">):</span>
            <span class="n">reference_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">yield</span> <span class="n">mutation_set</span><span class="p">,</span> <span class="n">reference_id</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="n">stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_statistics</span><span class="p">()</span>
        <span class="n">ref_count</span> <span class="o">=</span> <span class="n">stats</span><span class="p">[</span><span class="s2">&quot;num_reference_sequences&quot;</span><span class="p">]</span>
        <span class="n">ref_info</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot; (</span><span class="si">{</span><span class="n">ref_count</span><span class="si">}</span><span class="s2"> reference sequences)&quot;</span>
            <span class="k">if</span> <span class="n">ref_count</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="k">else</span> <span class="s2">&quot; (no references)&quot;</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;MutationDataset(</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">)</span><span class="si">{</span><span class="n">ref_info</span><span class="si">}</span><span class="s2">: &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">stats</span><span class="p">[</span><span class="s1">&#39;total_mutation_sets&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> mutation sets, &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">stats</span><span class="p">[</span><span class="s1">&#39;total_mutations&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> mutations&quot;</span>
        <span class="p">)</span>

<div class="viewcode-block" id="MutationDataset.add_reference_sequence">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.add_reference_sequence">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">add_reference_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="n">BaseSequence</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a reference sequence with a unique identifier&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">sequence_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Reference sequence with ID &#39;</span><span class="si">{</span><span class="n">sequence_id</span><span class="si">}</span><span class="s2">&#39; already exists&quot;</span>
            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">[</span><span class="n">sequence_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">sequence</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_df</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Reset cached DataFrame</span></div>


<div class="viewcode-block" id="MutationDataset.remove_reference_sequence">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.remove_reference_sequence">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">remove_reference_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Remove a reference sequence&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">sequence_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reference sequence with ID &#39;</span><span class="si">{</span><span class="n">sequence_id</span><span class="si">}</span><span class="s2">&#39; not found&quot;</span><span class="p">)</span>

        <span class="c1"># Check if any mutation sets reference this sequence</span>
        <span class="n">referencing_sets</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">idx</span>
            <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">ref_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">ref_id</span> <span class="o">==</span> <span class="n">sequence_id</span>
        <span class="p">]</span>
        <span class="k">if</span> <span class="n">referencing_sets</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Cannot remove sequence &#39;</span><span class="si">{</span><span class="n">sequence_id</span><span class="si">}</span><span class="s2">&#39; as it is referenced by &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">referencing_sets</span><span class="p">)</span><span class="si">}</span><span class="s2"> mutation sets. Remove the mutation sets first.&quot;</span>
            <span class="p">)</span>

        <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">[</span><span class="n">sequence_id</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_df</span> <span class="o">=</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="MutationDataset.get_reference_sequence">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.get_reference_sequence">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_reference_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">BaseSequence</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get a reference sequence by ID&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">sequence_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reference sequence with ID &#39;</span><span class="si">{</span><span class="n">sequence_id</span><span class="si">}</span><span class="s2">&#39; not found&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">[</span><span class="n">sequence_id</span><span class="p">]</span></div>


<div class="viewcode-block" id="MutationDataset.list_reference_sequences">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.list_reference_sequences">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">list_reference_sequences</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get list of all reference sequence IDs&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span></div>


<div class="viewcode-block" id="MutationDataset.add_mutation_set">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.add_mutation_set">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">add_mutation_set</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">mutation_set</span><span class="p">:</span> <span class="n">MutationSet</span><span class="p">,</span>
        <span class="n">reference_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">label</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a mutation set to the dataset, linking to a reference sequence&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">reference_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reference sequence with ID &#39;</span><span class="si">{</span><span class="n">reference_id</span><span class="si">}</span><span class="s2">&#39; not found&quot;</span><span class="p">)</span>

        <span class="n">mutation_set_index</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation_set</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span><span class="n">mutation_set_index</span><span class="p">]</span> <span class="o">=</span> <span class="n">reference_id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span><span class="p">[</span><span class="n">mutation_set_index</span><span class="p">]</span> <span class="o">=</span> <span class="n">label</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_df</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Reset cached DataFrame</span></div>


<div class="viewcode-block" id="MutationDataset.add_mutation_sets">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.add_mutation_sets">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">add_mutation_sets</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">mutation_sets</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="n">MutationSet</span><span class="p">],</span>
        <span class="n">reference_ids</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">labels</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Sequence</span><span class="p">[</span><span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add multiple mutation sets to the dataset&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference_ids</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutation_sets</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Number of reference_ids must match number of mutation_sets&quot;</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">labels</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutation_sets</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Number of labels must match number of mutation_sets&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">mutation_set</span><span class="p">,</span> <span class="n">ref_id</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">mutation_sets</span><span class="p">,</span> <span class="n">reference_ids</span><span class="p">)):</span>
            <span class="n">label</span> <span class="o">=</span> <span class="n">labels</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="n">labels</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_mutation_set</span><span class="p">(</span><span class="n">mutation_set</span><span class="p">,</span> <span class="n">ref_id</span><span class="p">,</span> <span class="n">label</span><span class="p">)</span></div>


<div class="viewcode-block" id="MutationDataset.set_mutation_set_reference">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.set_mutation_set_reference">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">set_mutation_set_reference</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mutation_set_index</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">reference_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set the reference sequence for a specific mutation set&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">mutation_set_index</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mutation set index </span><span class="si">{</span><span class="n">mutation_set_index</span><span class="si">}</span><span class="s2"> out of range&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">reference_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reference sequence with ID &#39;</span><span class="si">{</span><span class="n">reference_id</span><span class="si">}</span><span class="s2">&#39; not found&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span><span class="n">mutation_set_index</span><span class="p">]</span> <span class="o">=</span> <span class="n">reference_id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_df</span> <span class="o">=</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="MutationDataset.get_mutation_set_reference">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.get_mutation_set_reference">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_mutation_set_reference</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mutation_set_index</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the reference sequence ID for a specific mutation set&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">mutation_set_index</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mutation set index </span><span class="si">{</span><span class="n">mutation_set_index</span><span class="si">}</span><span class="s2"> out of range&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span><span class="n">mutation_set_index</span><span class="p">]</span></div>


<div class="viewcode-block" id="MutationDataset.set_mutation_set_label">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.set_mutation_set_label">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">set_mutation_set_label</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mutation_set_index</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">label</span><span class="p">:</span> <span class="nb">float</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set the label for a specific mutation set&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">mutation_set_index</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mutation set index </span><span class="si">{</span><span class="n">mutation_set_index</span><span class="si">}</span><span class="s2"> out of range&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span><span class="p">[</span><span class="n">mutation_set_index</span><span class="p">]</span> <span class="o">=</span> <span class="n">label</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_df</span> <span class="o">=</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="MutationDataset.get_mutation_set_label">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.get_mutation_set_label">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_mutation_set_label</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mutation_set_index</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Any</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the label for a specific mutation set&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">mutation_set_index</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mutation set index </span><span class="si">{</span><span class="n">mutation_set_index</span><span class="si">}</span><span class="s2"> out of range&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mutation_set_index</span><span class="p">)</span></div>


<div class="viewcode-block" id="MutationDataset.remove_mutation_set">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.remove_mutation_set">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">remove_mutation_set</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mutation_set_index</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Remove a mutation set from the dataset&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">mutation_set_index</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mutation set index </span><span class="si">{</span><span class="n">mutation_set_index</span><span class="si">}</span><span class="s2"> out of range&quot;</span><span class="p">)</span>

        <span class="c1"># Remove the mutation set</span>
        <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">[</span><span class="n">mutation_set_index</span><span class="p">]</span>

        <span class="c1"># Update the reference mapping (shift indices)</span>
        <span class="n">new_references</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">new_labels</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">ref_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">idx</span> <span class="o">&lt;</span> <span class="n">mutation_set_index</span><span class="p">:</span>
                <span class="n">new_references</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_id</span>
                <span class="n">new_labels</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">idx</span> <span class="o">&gt;</span> <span class="n">mutation_set_index</span><span class="p">:</span>
                <span class="n">new_references</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_id</span>
                <span class="n">new_labels</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="c1"># Skip the removed index</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span> <span class="o">=</span> <span class="n">new_references</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span> <span class="o">=</span> <span class="n">new_labels</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_df</span> <span class="o">=</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="MutationDataset.validate_against_references">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.validate_against_references">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">validate_against_references</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Validate mutations against their reference sequences&quot;&quot;&quot;</span>
        <span class="n">validation_results</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;valid_mutation_sets&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;invalid_mutation_sets&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;position_mismatches&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="p">}</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mutation_set</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">):</span>
            <span class="n">set_name</span> <span class="o">=</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">name</span> <span class="ow">or</span> <span class="sa">f</span><span class="s2">&quot;MutationSet_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span>

            <span class="c1"># All mutation sets must have a reference sequence</span>
            <span class="n">reference_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>  <span class="c1"># This should always exist</span>
            <span class="n">reference_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">[</span><span class="n">reference_id</span><span class="p">]</span>
            <span class="n">set_valid</span> <span class="o">=</span> <span class="kc">True</span>

            <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">mutations</span><span class="p">:</span>
                <span class="c1"># Check if mutation position is within sequence bounds</span>
                <span class="k">if</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">):</span>
                    <span class="n">validation_results</span><span class="p">[</span><span class="s2">&quot;invalid_mutation_sets&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="p">{</span>
                            <span class="s2">&quot;mutation_set&quot;</span><span class="p">:</span> <span class="n">set_name</span><span class="p">,</span>
                            <span class="s2">&quot;reference_id&quot;</span><span class="p">:</span> <span class="n">reference_id</span><span class="p">,</span>
                            <span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">mutation</span><span class="p">),</span>
                            <span class="s2">&quot;error&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;Position </span><span class="si">{</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="si">}</span><span class="s2"> exceeds sequence length (0-indexed)&quot;</span><span class="p">,</span>
                        <span class="p">}</span>
                    <span class="p">)</span>
                    <span class="n">set_valid</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">continue</span>

                <span class="c1"># Check if wild type matches reference for amino acid mutations</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">AminoAcidMutation</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span>
                    <span class="n">reference_sequence</span><span class="p">,</span> <span class="n">ProteinSequence</span>
                <span class="p">):</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">ref_residue</span> <span class="o">=</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">get_residue</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">ref_residue</span> <span class="o">!=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">wild_amino_acid</span><span class="p">:</span>
                            <span class="n">validation_results</span><span class="p">[</span><span class="s2">&quot;position_mismatches&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                                <span class="p">{</span>
                                    <span class="s2">&quot;mutation_set&quot;</span><span class="p">:</span> <span class="n">set_name</span><span class="p">,</span>
                                    <span class="s2">&quot;reference_id&quot;</span><span class="p">:</span> <span class="n">reference_id</span><span class="p">,</span>
                                    <span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">mutation</span><span class="p">),</span>
                                    <span class="s2">&quot;expected&quot;</span><span class="p">:</span> <span class="n">ref_residue</span><span class="p">,</span>
                                    <span class="s2">&quot;found&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">wild_amino_acid</span><span class="p">,</span>
                                    <span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="p">,</span>
                                <span class="p">}</span>
                            <span class="p">)</span>
                    <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
                        <span class="n">validation_results</span><span class="p">[</span><span class="s2">&quot;invalid_mutation_sets&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="p">{</span>
                                <span class="s2">&quot;mutation_set&quot;</span><span class="p">:</span> <span class="n">set_name</span><span class="p">,</span>
                                <span class="s2">&quot;reference_id&quot;</span><span class="p">:</span> <span class="n">reference_id</span><span class="p">,</span>
                                <span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">mutation</span><span class="p">),</span>
                                <span class="s2">&quot;error&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;Position </span><span class="si">{</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="si">}</span><span class="s2"> out of range&quot;</span><span class="p">,</span>
                            <span class="p">}</span>
                        <span class="p">)</span>
                        <span class="n">set_valid</span> <span class="o">=</span> <span class="kc">False</span>

                <span class="c1"># Check codon mutations for nucleotide sequences</span>
                <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">CodonMutation</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span>
                    <span class="n">reference_sequence</span><span class="p">,</span> <span class="p">(</span><span class="n">DNASequence</span><span class="p">,</span> <span class="n">RNASequence</span><span class="p">)</span>
                <span class="p">):</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="c1"># Assuming position is codon position, get the codon at this position</span>
                        <span class="n">start_pos</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">*</span> <span class="mi">3</span>
                        <span class="k">if</span> <span class="n">start_pos</span> <span class="o">+</span> <span class="mi">3</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">):</span>
                            <span class="n">ref_codon</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span>
                                <span class="n">reference_sequence</span><span class="p">[</span><span class="n">start_pos</span> <span class="p">:</span> <span class="n">start_pos</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]</span>
                            <span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                            <span class="k">if</span> <span class="n">ref_codon</span> <span class="o">!=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">wild_codon</span><span class="p">:</span>
                                <span class="n">validation_results</span><span class="p">[</span><span class="s2">&quot;position_mismatches&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                                    <span class="p">{</span>
                                        <span class="s2">&quot;mutation_set&quot;</span><span class="p">:</span> <span class="n">set_name</span><span class="p">,</span>
                                        <span class="s2">&quot;reference_id&quot;</span><span class="p">:</span> <span class="n">reference_id</span><span class="p">,</span>
                                        <span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">mutation</span><span class="p">),</span>
                                        <span class="s2">&quot;expected&quot;</span><span class="p">:</span> <span class="n">ref_codon</span><span class="p">,</span>
                                        <span class="s2">&quot;found&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">wild_codon</span><span class="p">,</span>
                                        <span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="p">,</span>
                                    <span class="p">}</span>
                                <span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">validation_results</span><span class="p">[</span><span class="s2">&quot;invalid_mutation_sets&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                                <span class="p">{</span>
                                    <span class="s2">&quot;mutation_set&quot;</span><span class="p">:</span> <span class="n">set_name</span><span class="p">,</span>
                                    <span class="s2">&quot;reference_id&quot;</span><span class="p">:</span> <span class="n">reference_id</span><span class="p">,</span>
                                    <span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">mutation</span><span class="p">),</span>
                                    <span class="s2">&quot;error&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;Codon position </span><span class="si">{</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="si">}</span><span class="s2"> exceeds sequence bounds&quot;</span><span class="p">,</span>
                                <span class="p">}</span>
                            <span class="p">)</span>
                            <span class="n">set_valid</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                        <span class="n">validation_results</span><span class="p">[</span><span class="s2">&quot;invalid_mutation_sets&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="p">{</span>
                                <span class="s2">&quot;mutation_set&quot;</span><span class="p">:</span> <span class="n">set_name</span><span class="p">,</span>
                                <span class="s2">&quot;reference_id&quot;</span><span class="p">:</span> <span class="n">reference_id</span><span class="p">,</span>
                                <span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">mutation</span><span class="p">),</span>
                                <span class="s2">&quot;error&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;Error validating codon: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                            <span class="p">}</span>
                        <span class="p">)</span>
                        <span class="n">set_valid</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="k">if</span> <span class="n">set_valid</span><span class="p">:</span>
                <span class="n">validation_results</span><span class="p">[</span><span class="s2">&quot;valid_mutation_sets&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;mutation_set&quot;</span><span class="p">:</span> <span class="n">set_name</span><span class="p">,</span>
                        <span class="s2">&quot;reference_id&quot;</span><span class="p">:</span> <span class="n">reference_id</span><span class="p">,</span>
                    <span class="p">}</span>
                <span class="p">)</span>

        <span class="k">return</span> <span class="n">validation_results</span></div>


<div class="viewcode-block" id="MutationDataset.to_dataframe">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.to_dataframe">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">to_dataframe</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Convert dataset to pandas DataFrame&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mutation_set</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
                <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Converting dataset to DataFrame: &quot;</span>
            <span class="p">):</span>
                <span class="n">reference_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span>
                    <span class="n">i</span>
                <span class="p">]</span>  <span class="c1"># This should always exist</span>
                <span class="n">reference_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">[</span><span class="n">reference_id</span><span class="p">]</span>
                <span class="n">label</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

                <span class="n">base_data</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="s2">&quot;mutation_set_id&quot;</span><span class="p">:</span> <span class="n">i</span><span class="p">,</span>
                    <span class="s2">&quot;mutation_set_name&quot;</span><span class="p">:</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                    <span class="s2">&quot;reference_id&quot;</span><span class="p">:</span> <span class="n">reference_id</span><span class="p">,</span>
                    <span class="s2">&quot;reference_sequence_name&quot;</span><span class="p">:</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                    <span class="s2">&quot;reference_sequence_length&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">),</span>
                    <span class="s2">&quot;reference_sequence_type&quot;</span><span class="p">:</span> <span class="nb">type</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span>
                    <span class="s2">&quot;num_mutations&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutation_set</span><span class="p">),</span>
                    <span class="s2">&quot;is_single_mutation&quot;</span><span class="p">:</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">is_single_mutation</span><span class="p">(),</span>
                    <span class="s2">&quot;is_valid&quot;</span><span class="p">:</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">validate_all</span><span class="p">(),</span>
                    <span class="s2">&quot;positions&quot;</span><span class="p">:</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">get_positions</span><span class="p">())),</span>
                    <span class="s2">&quot;mutation_subtype&quot;</span><span class="p">:</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">mutation_subtype</span><span class="p">,</span>
                    <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="n">label</span><span class="p">,</span>
                <span class="p">}</span>

                <span class="c1"># Add mutation set metadata</span>
                <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="n">base_data</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;set_</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>

                <span class="c1"># Add mutation information</span>
                <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mutation_set</span><span class="o">.</span><span class="n">mutations</span><span class="p">):</span>
                    <span class="n">mutation_data</span> <span class="o">=</span> <span class="n">base_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
                    <span class="n">mutation_data</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                        <span class="p">{</span>
                            <span class="s2">&quot;mutation_id&quot;</span><span class="p">:</span> <span class="n">j</span><span class="p">,</span>
                            <span class="s2">&quot;mutation_type&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">type</span><span class="p">,</span>
                            <span class="s2">&quot;mutation_string&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">mutation</span><span class="p">),</span>
                            <span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="p">,</span>
                            <span class="s2">&quot;mutation_category&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">get_mutation_category</span><span class="p">(),</span>
                        <span class="p">}</span>
                    <span class="p">)</span>

                    <span class="c1"># Add amino acid mutation-specific data</span>
                    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">AminoAcidMutation</span><span class="p">):</span>
                        <span class="n">mutation_data</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                            <span class="p">{</span>
                                <span class="s2">&quot;wild_amino_acid&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">wild_amino_acid</span><span class="p">,</span>
                                <span class="s2">&quot;mutant_amino_acid&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">mutant_amino_acid</span><span class="p">,</span>
                                <span class="s2">&quot;effect_type&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">effect_type</span><span class="p">,</span>
                                <span class="s2">&quot;is_synonymous&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">is_synonymous</span><span class="p">(),</span>
                                <span class="s2">&quot;is_nonsense&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">is_nonsense</span><span class="p">(),</span>
                                <span class="s2">&quot;is_missense&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">is_missense</span><span class="p">(),</span>
                            <span class="p">}</span>
                        <span class="p">)</span>

                        <span class="c1"># Add reference residue if available</span>
                        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">,</span> <span class="n">ProteinSequence</span><span class="p">):</span>
                            <span class="k">try</span><span class="p">:</span>
                                <span class="n">ref_residue</span> <span class="o">=</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">get_residue</span><span class="p">(</span>
                                    <span class="n">mutation</span><span class="o">.</span><span class="n">position</span>
                                <span class="p">)</span>
                                <span class="n">mutation_data</span><span class="p">[</span><span class="s2">&quot;reference_residue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_residue</span>
                                <span class="n">mutation_data</span><span class="p">[</span><span class="s2">&quot;wild_type_matches_reference&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                                    <span class="n">ref_residue</span> <span class="o">==</span> <span class="n">mutation</span><span class="o">.</span><span class="n">wild_amino_acid</span>
                                <span class="p">)</span>
                            <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
                                <span class="n">mutation_data</span><span class="p">[</span><span class="s2">&quot;reference_residue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
                                <span class="n">mutation_data</span><span class="p">[</span><span class="s2">&quot;wild_type_matches_reference&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>

                    <span class="c1"># Add codon mutation-specific data</span>
                    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">CodonMutation</span><span class="p">):</span>
                        <span class="n">mutation_data</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                            <span class="p">{</span>
                                <span class="s2">&quot;wild_codon&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">wild_codon</span><span class="p">,</span>
                                <span class="s2">&quot;mutant_codon&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">mutant_codon</span><span class="p">,</span>
                                <span class="s2">&quot;seq_type&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">seq_type</span><span class="p">,</span>
                            <span class="p">}</span>
                        <span class="p">)</span>

                        <span class="c1"># Add reference codon if available</span>
                        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">,</span> <span class="p">(</span><span class="n">DNASequence</span><span class="p">,</span> <span class="n">RNASequence</span><span class="p">)):</span>
                            <span class="k">try</span><span class="p">:</span>
                                <span class="c1"># Get the codon at this position (assuming position is codon position)</span>
                                <span class="n">start_pos</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">*</span> <span class="mi">3</span>
                                <span class="k">if</span> <span class="n">start_pos</span> <span class="o">+</span> <span class="mi">3</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">):</span>
                                    <span class="n">ref_codon</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span>
                                        <span class="n">reference_sequence</span><span class="p">[</span><span class="n">start_pos</span> <span class="p">:</span> <span class="n">start_pos</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]</span>
                                    <span class="p">)</span>
                                    <span class="n">mutation_data</span><span class="p">[</span><span class="s2">&quot;reference_codon&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_codon</span>
                                    <span class="n">mutation_data</span><span class="p">[</span><span class="s2">&quot;wild_codon_matches_reference&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                                        <span class="n">ref_codon</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">==</span> <span class="n">mutation</span><span class="o">.</span><span class="n">wild_codon</span>
                                    <span class="p">)</span>
                                <span class="k">else</span><span class="p">:</span>
                                    <span class="n">mutation_data</span><span class="p">[</span><span class="s2">&quot;reference_codon&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
                                    <span class="n">mutation_data</span><span class="p">[</span><span class="s2">&quot;wild_codon_matches_reference&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                                        <span class="kc">False</span>
                                    <span class="p">)</span>
                            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                                <span class="n">mutation_data</span><span class="p">[</span><span class="s2">&quot;reference_codon&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
                                <span class="n">mutation_data</span><span class="p">[</span><span class="s2">&quot;wild_codon_matches_reference&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>

                    <span class="c1"># Add mutation metadata</span>
                    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">mutation</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                        <span class="n">mutation_data</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;mutation_</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>

                    <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation_data</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_df</span></div>


<div class="viewcode-block" id="MutationDataset.filter_by_reference">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.filter_by_reference">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">filter_by_reference</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reference_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;MutationDataset&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Filter dataset to only include mutation sets from a specific reference sequence&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">reference_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reference sequence with ID &#39;</span><span class="si">{</span><span class="n">reference_id</span><span class="si">}</span><span class="s2">&#39; not found&quot;</span><span class="p">)</span>

        <span class="n">filtered_sets</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">filtered_references</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">filtered_labels</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mutation_set</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
            <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Filtering by reference: &quot;</span>
        <span class="p">):</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">reference_id</span><span class="p">:</span>
                <span class="n">filtered_sets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation_set</span><span class="p">)</span>
                <span class="n">filtered_references</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">reference_id</span><span class="p">)</span>
                <span class="n">filtered_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

        <span class="n">filtered_dataset</span> <span class="o">=</span> <span class="n">MutationDataset</span><span class="p">(</span>
            <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">reference_id</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="k">else</span> <span class="n">reference_id</span>
        <span class="p">)</span>
        <span class="n">filtered_dataset</span><span class="o">.</span><span class="n">add_reference_sequence</span><span class="p">(</span>
            <span class="n">reference_id</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">[</span><span class="n">reference_id</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">filtered_dataset</span><span class="o">.</span><span class="n">add_mutation_sets</span><span class="p">(</span>
            <span class="n">filtered_sets</span><span class="p">,</span> <span class="n">filtered_references</span><span class="p">,</span> <span class="n">filtered_labels</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">filtered_dataset</span></div>


<div class="viewcode-block" id="MutationDataset.filter_by_mutation_type">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.filter_by_mutation_type">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">filter_by_mutation_type</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">mutation_type</span><span class="p">:</span> <span class="n">Type</span><span class="p">[</span><span class="n">BaseMutation</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;MutationDataset&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Filter dataset by mutation type&quot;&quot;&quot;</span>
        <span class="n">filtered_sets</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">filtered_references</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">filtered_labels</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mutation_set</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
            <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Filtering by mutation type: &quot;</span>
        <span class="p">):</span>
            <span class="c1"># Filter mutations by type</span>
            <span class="n">filtered_mutations</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">m</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">mutations</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">mutation_type</span><span class="p">)</span>
            <span class="p">]</span>

            <span class="k">if</span> <span class="n">filtered_mutations</span><span class="p">:</span>
                <span class="c1"># Create new mutation set with filtered mutations</span>
                <span class="k">if</span> <span class="n">mutation_type</span> <span class="o">==</span> <span class="n">AminoAcidMutation</span><span class="p">:</span>
                    <span class="n">new_set</span> <span class="o">=</span> <span class="n">AminoAcidMutationSet</span><span class="p">(</span>
                        <span class="n">mutations</span><span class="o">=</span><span class="n">filtered_mutations</span><span class="p">,</span>  <span class="c1"># type: ignore</span>
                        <span class="n">name</span><span class="o">=</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">mutation_set</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_filtered&quot;</span>
                            <span class="k">if</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">name</span>
                            <span class="k">else</span> <span class="s2">&quot;filtered&quot;</span>
                        <span class="p">),</span>
                        <span class="n">metadata</span><span class="o">=</span><span class="n">mutation_set</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span>
                    <span class="p">)</span>
                <span class="k">elif</span> <span class="n">mutation_type</span> <span class="o">==</span> <span class="n">CodonMutation</span><span class="p">:</span>
                    <span class="n">new_set</span> <span class="o">=</span> <span class="n">CodonMutationSet</span><span class="p">(</span>
                        <span class="n">mutations</span><span class="o">=</span><span class="n">filtered_mutations</span><span class="p">,</span>  <span class="c1"># type: ignore</span>
                        <span class="n">name</span><span class="o">=</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">mutation_set</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_filtered&quot;</span>
                            <span class="k">if</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">name</span>
                            <span class="k">else</span> <span class="s2">&quot;filtered&quot;</span>
                        <span class="p">),</span>
                        <span class="n">metadata</span><span class="o">=</span><span class="n">mutation_set</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">new_set</span> <span class="o">=</span> <span class="n">MutationSet</span><span class="p">(</span>
                        <span class="n">mutations</span><span class="o">=</span><span class="n">filtered_mutations</span><span class="p">,</span>
                        <span class="n">mutation_type</span><span class="o">=</span><span class="n">mutation_type</span><span class="p">,</span>
                        <span class="n">name</span><span class="o">=</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">mutation_set</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_filtered&quot;</span>
                            <span class="k">if</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">name</span>
                            <span class="k">else</span> <span class="s2">&quot;filtered&quot;</span>
                        <span class="p">),</span>
                        <span class="n">metadata</span><span class="o">=</span><span class="n">mutation_set</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span>
                    <span class="p">)</span>

                <span class="n">filtered_sets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_set</span><span class="p">)</span>
                <span class="n">ref_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="n">filtered_references</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ref_id</span><span class="p">)</span>
                <span class="n">filtered_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

        <span class="n">filtered_dataset</span> <span class="o">=</span> <span class="n">MutationDataset</span><span class="p">(</span>
            <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_filtered&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="k">else</span> <span class="s2">&quot;filtered&quot;</span>
        <span class="p">)</span>
        <span class="c1"># Copy all reference sequences that are still needed</span>
        <span class="n">needed_refs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">filtered_references</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">ref_id</span> <span class="ow">in</span> <span class="n">needed_refs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ref_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">filtered_dataset</span><span class="o">.</span><span class="n">add_reference_sequence</span><span class="p">(</span>
                    <span class="n">ref_id</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">[</span><span class="n">ref_id</span><span class="p">]</span>
                <span class="p">)</span>

        <span class="n">filtered_dataset</span><span class="o">.</span><span class="n">add_mutation_sets</span><span class="p">(</span>
            <span class="n">filtered_sets</span><span class="p">,</span> <span class="n">filtered_references</span><span class="p">,</span> <span class="n">filtered_labels</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">filtered_dataset</span></div>


<div class="viewcode-block" id="MutationDataset.filter_by_effect_type">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.filter_by_effect_type">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">filter_by_effect_type</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">effect_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;MutationDataset&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Filter dataset by amino acid mutation effect type (synonymous, missense, nonsense)&quot;&quot;&quot;</span>
        <span class="n">filtered_sets</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">filtered_references</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">filtered_labels</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mutation_set</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
            <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Filtering by effect type: &quot;</span>
        <span class="p">):</span>
            <span class="c1"># Filter amino acid mutations by effect type</span>
            <span class="n">filtered_mutations</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">mutations</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">AminoAcidMutation</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">mutation</span><span class="o">.</span><span class="n">effect_type</span> <span class="o">==</span> <span class="n">effect_type</span><span class="p">:</span>
                        <span class="n">filtered_mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">filtered_mutations</span><span class="p">:</span>
                <span class="n">new_set</span> <span class="o">=</span> <span class="n">AminoAcidMutationSet</span><span class="p">(</span>
                    <span class="n">mutations</span><span class="o">=</span><span class="n">filtered_mutations</span><span class="p">,</span>
                    <span class="n">name</span><span class="o">=</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">mutation_set</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">effect_type</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="k">if</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">name</span>
                        <span class="k">else</span> <span class="n">effect_type</span>
                    <span class="p">),</span>
                    <span class="n">metadata</span><span class="o">=</span><span class="n">mutation_set</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span>
                <span class="p">)</span>
                <span class="n">filtered_sets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_set</span><span class="p">)</span>
                <span class="n">ref_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>  <span class="c1"># Always exists now</span>
                <span class="n">filtered_references</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ref_id</span><span class="p">)</span>
                <span class="n">filtered_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

        <span class="n">filtered_dataset</span> <span class="o">=</span> <span class="n">MutationDataset</span><span class="p">(</span>
            <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">effect_type</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="k">else</span> <span class="n">effect_type</span>
        <span class="p">)</span>
        <span class="c1"># Copy all reference sequences that are still needed</span>
        <span class="n">needed_refs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">filtered_references</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">ref_id</span> <span class="ow">in</span> <span class="n">needed_refs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ref_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">filtered_dataset</span><span class="o">.</span><span class="n">add_reference_sequence</span><span class="p">(</span>
                    <span class="n">ref_id</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">[</span><span class="n">ref_id</span><span class="p">]</span>
                <span class="p">)</span>

        <span class="n">filtered_dataset</span><span class="o">.</span><span class="n">add_mutation_sets</span><span class="p">(</span>
            <span class="n">filtered_sets</span><span class="p">,</span> <span class="n">filtered_references</span><span class="p">,</span> <span class="n">filtered_labels</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">filtered_dataset</span></div>


<div class="viewcode-block" id="MutationDataset.get_statistics">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.get_statistics">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_statistics</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get basic statistics about the dataset&quot;&quot;&quot;</span>
        <span class="n">total_sets</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">)</span>
        <span class="n">total_mutations</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ms</span><span class="p">)</span> <span class="k">for</span> <span class="n">ms</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">)</span>
        <span class="n">single_mutation_sets</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span>
            <span class="mi">1</span> <span class="k">for</span> <span class="n">ms</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span> <span class="k">if</span> <span class="n">ms</span><span class="o">.</span><span class="n">is_single_mutation</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">multiple_mutation_sets</span> <span class="o">=</span> <span class="n">total_sets</span> <span class="o">-</span> <span class="n">single_mutation_sets</span>

        <span class="n">mutation_types</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">mutation_categories</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">effect_types</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">reference_stats</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># Statistics by reference sequence</span>
        <span class="k">for</span> <span class="n">ref_id</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Statistics - ref seq: &quot;</span>
        <span class="p">):</span>
            <span class="n">reference_stats</span><span class="p">[</span><span class="n">ref_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;sequence_name&quot;</span><span class="p">:</span> <span class="n">sequence</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                <span class="s2">&quot;sequence_length&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span>
                <span class="s2">&quot;sequence_type&quot;</span><span class="p">:</span> <span class="nb">type</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span>
                <span class="s2">&quot;mutation_sets&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                <span class="s2">&quot;mutations&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="p">}</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mutation_set</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
            <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Statistics - mutation sets: &quot;</span>
        <span class="p">):</span>
            <span class="n">ref_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>  <span class="c1"># Always exists now</span>
            <span class="k">if</span> <span class="n">ref_id</span> <span class="ow">in</span> <span class="n">reference_stats</span><span class="p">:</span>
                <span class="n">reference_stats</span><span class="p">[</span><span class="n">ref_id</span><span class="p">][</span><span class="s2">&quot;mutation_sets&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">reference_stats</span><span class="p">[</span><span class="n">ref_id</span><span class="p">][</span><span class="s2">&quot;mutations&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutation_set</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">mutations</span><span class="p">:</span>
                <span class="c1"># Count mutation types</span>
                <span class="n">mut_type</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">type</span>
                <span class="n">mutation_types</span><span class="p">[</span><span class="n">mut_type</span><span class="p">]</span> <span class="o">=</span> <span class="n">mutation_types</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mut_type</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>

                <span class="c1"># Count mutation categories</span>
                <span class="n">category</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">get_mutation_category</span><span class="p">()</span>
                <span class="n">mutation_categories</span><span class="p">[</span><span class="n">category</span><span class="p">]</span> <span class="o">=</span> <span class="n">mutation_categories</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">category</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>

                <span class="c1"># Count effect types for amino acid mutations</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">AminoAcidMutation</span><span class="p">):</span>
                    <span class="n">effect</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">effect_type</span>
                    <span class="n">effect_types</span><span class="p">[</span><span class="n">effect</span><span class="p">]</span> <span class="o">=</span> <span class="n">effect_types</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">effect</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>

        <span class="n">stats</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;total_mutation_sets&quot;</span><span class="p">:</span> <span class="n">total_sets</span><span class="p">,</span>
            <span class="s2">&quot;total_mutations&quot;</span><span class="p">:</span> <span class="n">total_mutations</span><span class="p">,</span>
            <span class="s2">&quot;single_mutation_sets&quot;</span><span class="p">:</span> <span class="n">single_mutation_sets</span><span class="p">,</span>
            <span class="s2">&quot;multiple_mutation_sets&quot;</span><span class="p">:</span> <span class="n">multiple_mutation_sets</span><span class="p">,</span>
            <span class="s2">&quot;mutation_types&quot;</span><span class="p">:</span> <span class="n">mutation_types</span><span class="p">,</span>
            <span class="s2">&quot;mutation_categories&quot;</span><span class="p">:</span> <span class="n">mutation_categories</span><span class="p">,</span>
            <span class="s2">&quot;effect_types&quot;</span><span class="p">:</span> <span class="n">effect_types</span><span class="p">,</span>
            <span class="s2">&quot;average_mutations_per_set&quot;</span><span class="p">:</span> <span class="p">(</span>
                <span class="n">total_mutations</span> <span class="o">/</span> <span class="n">total_sets</span> <span class="k">if</span> <span class="n">total_sets</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
            <span class="p">),</span>
            <span class="s2">&quot;reference_sequences&quot;</span><span class="p">:</span> <span class="n">reference_stats</span><span class="p">,</span>
            <span class="s2">&quot;num_reference_sequences&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">),</span>
        <span class="p">}</span>

        <span class="k">return</span> <span class="n">stats</span></div>


<div class="viewcode-block" id="MutationDataset.get_position_coverage">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.get_position_coverage">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_position_coverage</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">reference_id</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get statistics about position coverage across reference sequences&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">reference_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">reference_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Reference sequence with ID &#39;</span><span class="si">{</span><span class="n">reference_id</span><span class="si">}</span><span class="s2">&#39; not found&quot;</span>
                <span class="p">)</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_single_sequence_coverage</span><span class="p">(</span><span class="n">reference_id</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Get coverage for all sequences</span>
            <span class="n">coverage_stats</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">ref_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">:</span>
                <span class="n">coverage_stats</span><span class="p">[</span><span class="n">ref_id</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_single_sequence_coverage</span><span class="p">(</span><span class="n">ref_id</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">coverage_stats</span></div>


    <span class="k">def</span><span class="w"> </span><span class="nf">_get_single_sequence_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reference_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get position coverage for a single reference sequence&quot;&quot;&quot;</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">[</span><span class="n">reference_id</span><span class="p">]</span>
        <span class="n">all_positions</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mutation_set</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">):</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">reference_id</span><span class="p">:</span>  <span class="c1"># Always exists now</span>
                <span class="n">all_positions</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">mutation_set</span><span class="o">.</span><span class="n">get_positions</span><span class="p">())</span>

        <span class="n">seq_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="n">covered_positions</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_positions</span><span class="p">)</span>
        <span class="n">coverage_percentage</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span><span class="n">covered_positions</span> <span class="o">/</span> <span class="n">seq_length</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span> <span class="k">if</span> <span class="n">seq_length</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span>
            <span class="s2">&quot;reference_id&quot;</span><span class="p">:</span> <span class="n">reference_id</span><span class="p">,</span>
            <span class="s2">&quot;sequence_name&quot;</span><span class="p">:</span> <span class="n">sequence</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="s2">&quot;sequence_length&quot;</span><span class="p">:</span> <span class="n">seq_length</span><span class="p">,</span>
            <span class="s2">&quot;sequence_type&quot;</span><span class="p">:</span> <span class="nb">type</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span>
            <span class="s2">&quot;covered_positions&quot;</span><span class="p">:</span> <span class="n">covered_positions</span><span class="p">,</span>
            <span class="s2">&quot;uncovered_positions&quot;</span><span class="p">:</span> <span class="n">seq_length</span> <span class="o">-</span> <span class="n">covered_positions</span><span class="p">,</span>
            <span class="s2">&quot;coverage_percentage&quot;</span><span class="p">:</span> <span class="n">coverage_percentage</span><span class="p">,</span>
            <span class="s2">&quot;position_list&quot;</span><span class="p">:</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">all_positions</span><span class="p">)),</span>
        <span class="p">}</span>

<div class="viewcode-block" id="MutationDataset.convert_codon_to_amino_acid_sets">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.convert_codon_to_amino_acid_sets">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">convert_codon_to_amino_acid_sets</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">convert_labels</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;MutationDataset&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Convert all codon mutation sets to amino acid mutation sets</span>

<span class="sd">        Parameters:</span>
<span class="sd">            convert_labels: Whether to save the labels with the mutation sets (default: False)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">converted_sets</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">converted_references</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">converted_labels</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mutation_set</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation_set</span><span class="p">,</span> <span class="n">CodonMutationSet</span><span class="p">):</span>
                <span class="n">aa_set</span> <span class="o">=</span> <span class="n">mutation_set</span><span class="o">.</span><span class="n">to_amino_acid_mutation_set</span><span class="p">()</span>
                <span class="n">converted_sets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aa_set</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">converted_sets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation_set</span><span class="p">)</span>

            <span class="n">ref_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">converted_references</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ref_id</span><span class="p">)</span>
            <span class="n">converted_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_mutation_set_label</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

        <span class="n">converted_dataset</span> <span class="o">=</span> <span class="n">MutationDataset</span><span class="p">(</span>
            <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_aa_converted&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="k">else</span> <span class="s2">&quot;aa_converted&quot;</span>
        <span class="p">)</span>

        <span class="c1"># Copy all reference sequences</span>
        <span class="k">for</span> <span class="n">ref_id</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">converted_dataset</span><span class="o">.</span><span class="n">add_reference_sequence</span><span class="p">(</span><span class="n">ref_id</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">convert_labels</span><span class="p">:</span>
            <span class="n">converted_dataset</span><span class="o">.</span><span class="n">add_mutation_sets</span><span class="p">(</span><span class="n">converted_sets</span><span class="p">,</span> <span class="n">converted_references</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">converted_dataset</span><span class="o">.</span><span class="n">add_mutation_sets</span><span class="p">(</span>
                <span class="n">converted_sets</span><span class="p">,</span> <span class="n">converted_references</span><span class="p">,</span> <span class="n">converted_labels</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">converted_dataset</span></div>


    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_sanitize_filename</span><span class="p">(</span><span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Sanitize a string to be used as a filename or directory name&quot;&quot;&quot;</span>
        <span class="kn">import</span><span class="w"> </span><span class="nn">re</span>

        <span class="c1"># Replace invalid characters with underscores</span>
        <span class="n">sanitized</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;[&lt;&gt;:&quot;/</span><span class="se">\\</span><span class="s1">|?*]&#39;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>
        <span class="c1"># Remove leading/trailing whitespace and dots</span>
        <span class="n">sanitized</span> <span class="o">=</span> <span class="n">sanitized</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s2">&quot;. &quot;</span><span class="p">)</span>
        <span class="c1"># Ensure it&#39;s not empty</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">sanitized</span><span class="p">:</span>
            <span class="n">sanitized</span> <span class="o">=</span> <span class="s2">&quot;unnamed&quot;</span>
        <span class="c1"># Limit length to avoid filesystem issues</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sanitized</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">200</span><span class="p">:</span>
            <span class="n">sanitized</span> <span class="o">=</span> <span class="n">sanitized</span><span class="p">[:</span><span class="mi">200</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">sanitized</span>

<div class="viewcode-block" id="MutationDataset.save_by_reference">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.save_by_reference">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">save_by_reference</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">base_dir</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Save dataset by reference_id, creating separate folders for each reference.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            base_dir: Base directory to create reference folders in</span>

<span class="sd">        For each reference_id, creates:</span>
<span class="sd">            - {base_dir}/{reference_id}/data.csv: mutation data with columns [mutation_name, mutated_sequence, label]</span>
<span class="sd">            - {base_dir}/{reference_id}/wt.fasta: wild-type reference sequence</span>
<span class="sd">            - {base_dir}/{reference_id}/metadata.json: statistics and metadata for this reference</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">base_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">base_dir</span><span class="p">)</span>
        <span class="n">base_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saving dataset by reference to: </span><span class="si">{</span><span class="n">base_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Group mutation sets by reference_id</span>
        <span class="n">ref_groups</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mutation_set</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
            <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Grouping mutation sets by reference_id&quot;</span>
        <span class="p">):</span>
            <span class="n">ref_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_references</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">ref_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ref_groups</span><span class="p">:</span>
                <span class="n">ref_groups</span><span class="p">[</span><span class="n">ref_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">ref_groups</span><span class="p">[</span><span class="n">ref_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">i</span><span class="p">,</span> <span class="n">mutation_set</span><span class="p">))</span>

        <span class="c1"># Process each reference</span>
        <span class="k">for</span> <span class="n">ref_id</span><span class="p">,</span> <span class="n">mutation_set_list</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
            <span class="n">ref_groups</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Processing references&quot;</span>
        <span class="p">):</span>
            <span class="c1"># Create sanitized directory name</span>
            <span class="n">sanitized_ref_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sanitize_filename</span><span class="p">(</span><span class="n">ref_id</span><span class="p">)</span>
            <span class="n">ref_dir</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="n">sanitized_ref_id</span>
            <span class="n">ref_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

            <span class="c1"># Get reference sequence</span>
            <span class="n">ref_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">[</span><span class="n">ref_id</span><span class="p">]</span>

            <span class="c1"># Prepare data for CSV</span>
            <span class="n">csv_data</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="k">for</span> <span class="n">set_index</span><span class="p">,</span> <span class="n">mutation_set</span> <span class="ow">in</span> <span class="n">mutation_set_list</span><span class="p">:</span>
                <span class="c1"># Get mutation name (string representation of mutation set)</span>
                <span class="n">mutation_name</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">mutation_set</span><span class="p">)</span>

                <span class="c1"># Apply mutations to get mutated sequence</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">mutated_sequence</span> <span class="o">=</span> <span class="n">ref_sequence</span><span class="o">.</span><span class="n">apply_mutation</span><span class="p">(</span><span class="n">mutation_set</span><span class="p">)</span>
                    <span class="n">mutated_seq_str</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">mutated_sequence</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Warning: Could not apply mutations for </span><span class="si">{</span><span class="n">mutation_name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                    <span class="n">mutated_seq_str</span> <span class="o">=</span> <span class="s2">&quot;ERROR_APPLYING_MUTATION&quot;</span>

                <span class="c1"># Get label</span>
                <span class="n">label</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">set_index</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>

                <span class="n">csv_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;mutation_name&quot;</span><span class="p">:</span> <span class="n">mutation_name</span><span class="p">,</span>
                        <span class="s2">&quot;mutated_sequence&quot;</span><span class="p">:</span> <span class="n">mutated_seq_str</span><span class="p">,</span>
                        <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="n">label</span><span class="p">,</span>
                    <span class="p">}</span>
                <span class="p">)</span>

            <span class="c1"># Save data.csv</span>
            <span class="n">csv_path</span> <span class="o">=</span> <span class="n">ref_dir</span> <span class="o">/</span> <span class="s2">&quot;data.csv&quot;</span>
            <span class="n">df_ref</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">csv_data</span><span class="p">)</span>
            <span class="n">df_ref</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">csv_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

            <span class="c1"># Save wt.fasta</span>
            <span class="n">fasta_path</span> <span class="o">=</span> <span class="n">ref_dir</span> <span class="o">/</span> <span class="s2">&quot;wt.fasta&quot;</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="c1"># Use sequence name if available, otherwise use ref_id</span>
                <span class="n">seq_name</span> <span class="o">=</span> <span class="n">ref_sequence</span><span class="o">.</span><span class="n">name</span> <span class="k">if</span> <span class="n">ref_sequence</span><span class="o">.</span><span class="n">name</span> <span class="k">else</span> <span class="n">ref_id</span>
                <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">seq_name</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">ref_sequence</span><span class="p">)</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># Prepare metadata</span>
            <span class="n">ref_stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_single_sequence_coverage</span><span class="p">(</span><span class="n">ref_id</span><span class="p">)</span>
            <span class="n">metadata</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;reference_id&quot;</span><span class="p">:</span> <span class="n">ref_id</span><span class="p">,</span>
                <span class="s2">&quot;sequence_name&quot;</span><span class="p">:</span> <span class="n">ref_sequence</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                <span class="s2">&quot;sequence_type&quot;</span><span class="p">:</span> <span class="nb">type</span><span class="p">(</span><span class="n">ref_sequence</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span>
                <span class="s2">&quot;sequence_length&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_sequence</span><span class="p">),</span>
                <span class="s2">&quot;num_mutation_sets&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutation_set_list</span><span class="p">),</span>
                <span class="s2">&quot;total_mutations&quot;</span><span class="p">:</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ms</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">ms</span> <span class="ow">in</span> <span class="n">mutation_set_list</span><span class="p">),</span>
                <span class="s2">&quot;coverage_stats&quot;</span><span class="p">:</span> <span class="n">ref_stats</span><span class="p">,</span>
                <span class="s2">&quot;dataset_name&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                <span class="s2">&quot;sanitized_directory_name&quot;</span><span class="p">:</span> <span class="n">sanitized_ref_id</span><span class="p">,</span>
            <span class="p">}</span>

            <span class="c1"># Add label distribution</span>
            <span class="n">label_counts</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">set_index</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">mutation_set_list</span><span class="p">:</span>
                <span class="n">label</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_set_labels</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">set_index</span><span class="p">,</span> <span class="s2">&quot;unlabeled&quot;</span><span class="p">)</span>
                <span class="n">label_counts</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">label</span><span class="p">)]</span> <span class="o">=</span> <span class="n">label_counts</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">label</span><span class="p">),</span> <span class="mi">0</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;label_distribution&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">label_counts</span>

            <span class="c1"># Save metadata.json</span>
            <span class="n">metadata_path</span> <span class="o">=</span> <span class="n">ref_dir</span> <span class="o">/</span> <span class="s2">&quot;metadata.json&quot;</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">metadata_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="nb">str</span><span class="p">)</span></div>


<div class="viewcode-block" id="MutationDataset.save">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.save">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">save</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">save_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;tidymut&quot;</span><span class="p">,</span> <span class="s2">&quot;pickle&quot;</span><span class="p">,</span> <span class="s2">&quot;dataframe&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="s2">&quot;tidymut&quot;</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Save the dataset to files.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            filepath: Base filepath (without extension)</span>
<span class="sd">            save_type: Type of save format (&quot;tidymut&quot;, &quot;dataframe&quot; or &quot;pickle&quot;)</span>

<span class="sd">        For save_type=&quot;dataframe&quot;:</span>
<span class="sd">            - Saves mutations as {filepath}.csv</span>
<span class="sd">            - Saves reference sequences as {filepath}_refs.pkl</span>
<span class="sd">            - Saves metadata as {filepath}_meta.json</span>

<span class="sd">        For save_type=&quot;pickle&quot;:</span>
<span class="sd">            - Saves entire dataset as {filepath}.pkl</span>

<span class="sd">        Example:</span>
<span class="sd">            dataset.save(&quot;my_study&quot;, &quot;dataframe&quot;)</span>
<span class="sd">            # Creates: my_study.csv, my_study_refs.pkl, my_study_meta.json</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">base_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">save_type</span> <span class="o">==</span> <span class="s2">&quot;dataframe&quot;</span><span class="p">:</span>
            <span class="c1"># Save mutations as CSV</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
            <span class="n">csv_path</span> <span class="o">=</span> <span class="n">base_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;.csv&quot;</span><span class="p">)</span>
            <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">csv_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

            <span class="c1"># Save reference sequences as pickle</span>
            <span class="n">refs_path</span> <span class="o">=</span> <span class="n">base_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_path</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_refs.pkl&quot;</span>
            <span class="p">)</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">refs_path</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>

            <span class="c1"># Save dataset metadata as JSON</span>
            <span class="n">meta_path</span> <span class="o">=</span> <span class="n">base_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_path</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_meta.json&quot;</span>
            <span class="p">)</span>
            <span class="n">dataset_meta</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                <span class="s2">&quot;metadata&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
                <span class="s2">&quot;save_type&quot;</span><span class="p">:</span> <span class="n">save_type</span><span class="p">,</span>
                <span class="s2">&quot;num_mutation_sets&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_sets</span><span class="p">),</span>
                <span class="s2">&quot;num_reference_sequences&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequences</span><span class="p">),</span>
            <span class="p">}</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">meta_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">dataset_meta</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

            <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Dataset saved to:&quot;</span><span class="p">)</span>
            <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Mutations: </span><span class="si">{</span><span class="n">csv_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  References: </span><span class="si">{</span><span class="n">refs_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Metadata: </span><span class="si">{</span><span class="n">meta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">elif</span> <span class="n">save_type</span> <span class="o">==</span> <span class="s2">&quot;pickle&quot;</span><span class="p">:</span>
            <span class="c1"># Save entire dataset as pickle</span>
            <span class="n">pkl_path</span> <span class="o">=</span> <span class="n">base_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;.pkl&quot;</span><span class="p">)</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">pkl_path</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
            <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Dataset saved to: </span><span class="si">{</span><span class="n">pkl_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">elif</span> <span class="n">save_type</span> <span class="o">==</span> <span class="s2">&quot;tidymut&quot;</span><span class="p">:</span>
            <span class="c1"># Save as TidyMut format</span>
            <span class="k">if</span> <span class="n">base_path</span><span class="o">.</span><span class="n">suffix</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Invalid TidyMut save format. Expected folder but got </span><span class="si">{</span><span class="n">base_path</span><span class="o">.</span><span class="n">suffix</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">save_by_reference</span><span class="p">(</span><span class="n">base_path</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Unsupported save_type: </span><span class="si">{</span><span class="n">save_type</span><span class="si">}</span><span class="s2">. Use &#39;tidymut&#39;, &#39;dataframe&#39; or &#39;pickle&#39;&quot;</span>
            <span class="p">)</span></div>


    <span class="c1"># ====== load ======</span>
<div class="viewcode-block" id="MutationDataset.load_by_reference">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.load_by_reference">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">load_by_reference</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">,</span>
        <span class="n">base_dir</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">],</span>
        <span class="n">dataset_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">is_zero_based</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;MutationDataset&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Load a dataset from tidymut reference-based format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        base_dir : Union[str, Path]</span>
<span class="sd">            Base directory containing reference folders</span>
<span class="sd">        dataset_name : Optional[str], default=None</span>
<span class="sd">            Optional name for the loaded dataset</span>
<span class="sd">        is_zero_based : bool, default=True</span>
<span class="sd">            Whether origin mutation positions are zero-based</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        MutationDataset instance</span>

<span class="sd">        Expected directory structure:</span>
<span class="sd">            base_dir/</span>
<span class="sd">            ├── reference_id_1/</span>
<span class="sd">            │   ├── data.csv</span>
<span class="sd">            │   ├── wt.fasta</span>
<span class="sd">            │   └── metadata.json</span>
<span class="sd">            ├── reference_id_2/</span>
<span class="sd">            │   ├── data.csv</span>
<span class="sd">            │   ├── wt.fasta</span>
<span class="sd">            │   └── metadata.json</span>
<span class="sd">            └── ...</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">import</span><span class="w"> </span><span class="nn">json</span>

        <span class="n">base_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">base_dir</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">base_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Base directory not found: </span><span class="si">{</span><span class="n">base_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">base_path</span><span class="o">.</span><span class="n">is_dir</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Path is not a directory: </span><span class="si">{</span><span class="n">base_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loading dataset from: </span><span class="si">{</span><span class="n">base_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Find all reference directories</span>
        <span class="n">ref_dirs</span> <span class="o">=</span> <span class="p">[</span><span class="n">d</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">base_path</span><span class="o">.</span><span class="n">iterdir</span><span class="p">()</span> <span class="k">if</span> <span class="n">d</span><span class="o">.</span><span class="n">is_dir</span><span class="p">()]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">ref_dirs</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No reference directories found in </span><span class="si">{</span><span class="n">base_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Create new dataset</span>
        <span class="n">dataset</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">dataset_name</span><span class="p">)</span>

        <span class="c1"># Process each reference directory</span>
        <span class="k">for</span> <span class="n">ref_dir</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">ref_dirs</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Loading references&quot;</span><span class="p">):</span>
            <span class="c1"># Required files</span>
            <span class="n">data_path</span> <span class="o">=</span> <span class="n">ref_dir</span> <span class="o">/</span> <span class="s2">&quot;data.csv&quot;</span>
            <span class="n">fasta_path</span> <span class="o">=</span> <span class="n">ref_dir</span> <span class="o">/</span> <span class="s2">&quot;wt.fasta&quot;</span>
            <span class="n">metadata_path</span> <span class="o">=</span> <span class="n">ref_dir</span> <span class="o">/</span> <span class="s2">&quot;metadata.json&quot;</span>

            <span class="c1"># Check required files exist</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">data_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Skipping </span><span class="si">{</span><span class="n">ref_dir</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2"> - data.csv not found&quot;</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">fasta_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Skipping </span><span class="si">{</span><span class="n">ref_dir</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2"> - wt.fasta not found&quot;</span><span class="p">)</span>
                <span class="k">continue</span>

            <span class="c1"># Load metadata to get original reference_id</span>
            <span class="n">original_ref_id</span> <span class="o">=</span> <span class="n">ref_dir</span><span class="o">.</span><span class="n">name</span>  <span class="c1"># Default to directory name</span>
            <span class="k">if</span> <span class="n">metadata_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">metadata_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                        <span class="n">metadata</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
                    <span class="n">original_ref_id</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;reference_id&quot;</span><span class="p">,</span> <span class="n">ref_dir</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Warning: Could not load metadata for </span><span class="si">{</span><span class="n">ref_dir</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>

            <span class="c1"># Load reference sequence from FASTA</span>
            <span class="n">sequence_type</span> <span class="o">=</span> <span class="n">SEQUENCE_TYPE_MAP</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
                <span class="n">metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;sequence_type&quot;</span><span class="p">,</span> <span class="s2">&quot;ProteinSequence&quot;</span><span class="p">),</span> <span class="n">ProteinSequence</span>
            <span class="p">)</span>

            <span class="n">ref_sequence</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
                <span class="n">load_sequences_from_fasta</span><span class="p">(</span>
                    <span class="n">fasta_path</span><span class="p">,</span> <span class="n">sequence_type</span><span class="p">,</span> <span class="n">header_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
            <span class="p">)[</span>
                <span class="mi">0</span>
            <span class="p">]</span>  <span class="c1"># get first sequence when mutli sequences</span>
            <span class="n">dataset</span><span class="o">.</span><span class="n">add_reference_sequence</span><span class="p">(</span><span class="n">original_ref_id</span><span class="p">,</span> <span class="n">ref_sequence</span><span class="p">)</span>

            <span class="c1"># Load mutation data</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">df_ref</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">data_path</span><span class="p">)</span>
                <span class="n">required_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;mutation_name&quot;</span><span class="p">,</span> <span class="s2">&quot;mutated_sequence&quot;</span><span class="p">,</span> <span class="s2">&quot;label&quot;</span><span class="p">]</span>
                <span class="n">missing_cols</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">required_cols</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df_ref</span><span class="o">.</span><span class="n">columns</span>
                <span class="p">]</span>
                <span class="k">if</span> <span class="n">missing_cols</span><span class="p">:</span>
                    <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Warning: Skipping </span><span class="si">{</span><span class="n">ref_dir</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2"> - missing columns: </span><span class="si">{</span><span class="n">missing_cols</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                    <span class="k">continue</span>

                <span class="c1"># Process each mutation set</span>
                <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_ref</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                    <span class="n">mutation_name</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;mutation_name&quot;</span><span class="p">]</span>
                    <span class="n">label</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span>

                    <span class="c1"># Parse mutation from mutation_name</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">mutation_set</span> <span class="o">=</span> <span class="n">MutationSet</span><span class="o">.</span><span class="n">from_string</span><span class="p">(</span>
                            <span class="n">mutation_name</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">is_zero_based</span><span class="o">=</span><span class="n">is_zero_based</span>
                        <span class="p">)</span>
                        <span class="n">dataset</span><span class="o">.</span><span class="n">add_mutation_set</span><span class="p">(</span><span class="n">mutation_set</span><span class="p">,</span> <span class="n">original_ref_id</span><span class="p">,</span> <span class="n">label</span><span class="p">)</span>
                    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                        <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Warning: Could not parse mutation &#39;</span><span class="si">{</span><span class="n">mutation_name</span><span class="si">}</span><span class="s2">&#39;: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">)</span>
                        <span class="k">continue</span>

                <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Loaded </span><span class="si">{</span><span class="n">original_ref_id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">df_ref</span><span class="p">)</span><span class="si">}</span><span class="s2"> mutation sets&quot;</span><span class="p">)</span>

            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Could not load data for </span><span class="si">{</span><span class="n">ref_dir</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">continue</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No valid mutation sets were loaded&quot;</span><span class="p">)</span>

        <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Successfully loaded dataset with </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span><span class="si">}</span><span class="s2"> mutation sets&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">dataset</span></div>


<div class="viewcode-block" id="MutationDataset.load">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.load">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">load</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">load_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;MutationDataset&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Load a dataset from files.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filepath : str</span>
<span class="sd">            Base filepath (with or without extension)</span>
<span class="sd">        load_type : Optional[str], default=None</span>
<span class="sd">            Type of load format (&quot;tidymut&quot;, &quot;dataframe&quot; or &quot;pickle&quot;).</span>
<span class="sd">            If None, auto-detect from file extension.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            MutationDataset instance</span>

<span class="sd">        Example</span>
<span class="sd">        -------</span>
<span class="sd">        &gt;&gt;&gt; # Auto-detect from extension</span>
<span class="sd">        &gt;&gt;&gt; dataset = MutationDataset.load(&quot;my_study.csv&quot;)</span>
<span class="sd">        &gt;&gt;&gt; dataset = MutationDataset.load(&quot;my_study.pkl&quot;)</span>

<span class="sd">        &gt;&gt;&gt; # Explicit type</span>
<span class="sd">        &gt;&gt;&gt; dataset = MutationDataset.load(&quot;my_study&quot;, &quot;dataframe&quot;)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">base_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>

        <span class="c1"># Auto-detect load type from extension if not specified</span>
        <span class="k">if</span> <span class="n">load_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">base_path</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.csv&quot;</span><span class="p">:</span>
                <span class="n">load_type</span> <span class="o">=</span> <span class="s2">&quot;dataframe&quot;</span>
            <span class="k">elif</span> <span class="n">base_path</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.pkl&quot;</span><span class="p">:</span>
                <span class="n">load_type</span> <span class="o">=</span> <span class="s2">&quot;pickle&quot;</span>
            <span class="k">elif</span> <span class="n">base_path</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">load_type</span> <span class="o">=</span> <span class="s2">&quot;tidymut&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Try dataframe format first</span>
                <span class="n">load_type</span> <span class="o">=</span> <span class="s2">&quot;dataframe&quot;</span>
                <span class="n">base_path</span> <span class="o">=</span> <span class="n">base_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>  <span class="c1"># Remove any extension</span>

        <span class="k">if</span> <span class="n">load_type</span> <span class="o">==</span> <span class="s2">&quot;dataframe&quot;</span><span class="p">:</span>
            <span class="c1"># Remove extension to get base path</span>
            <span class="k">if</span> <span class="n">base_path</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.csv&quot;</span><span class="p">:</span>
                <span class="n">base_path</span> <span class="o">=</span> <span class="n">base_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

            <span class="n">csv_path</span> <span class="o">=</span> <span class="n">base_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;.csv&quot;</span><span class="p">)</span>
            <span class="n">refs_path</span> <span class="o">=</span> <span class="n">base_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_path</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_refs.pkl&quot;</span>
            <span class="p">)</span>
            <span class="n">meta_path</span> <span class="o">=</span> <span class="n">base_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_path</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_meta.json&quot;</span>
            <span class="p">)</span>

            <span class="c1"># Check if files exist</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">csv_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;CSV file not found: </span><span class="si">{</span><span class="n">csv_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">refs_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;References file not found: </span><span class="si">{</span><span class="n">refs_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># Load mutations DataFrame</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">csv_path</span><span class="p">)</span>

            <span class="c1"># Load reference sequences</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">refs_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">reference_sequences</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

            <span class="c1"># Load metadata if available</span>
            <span class="n">dataset_name</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">dataset_metadata</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">if</span> <span class="n">meta_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">meta_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">meta</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
                    <span class="n">dataset_name</span> <span class="o">=</span> <span class="n">meta</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">)</span>
                    <span class="n">dataset_metadata</span> <span class="o">=</span> <span class="n">meta</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;metadata&quot;</span><span class="p">,</span> <span class="p">{})</span>

            <span class="c1"># Create dataset using `from_dataframe`</span>
            <span class="n">dataset</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">from_dataframe</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">reference_sequences</span><span class="p">,</span> <span class="n">dataset_name</span><span class="p">)</span>
            <span class="n">dataset</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">dataset_metadata</span>

            <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Dataset loaded from:&quot;</span><span class="p">)</span>
            <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Mutations: </span><span class="si">{</span><span class="n">csv_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  References: </span><span class="si">{</span><span class="n">refs_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">meta_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Metadata: </span><span class="si">{</span><span class="n">meta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="k">return</span> <span class="n">dataset</span>

        <span class="k">elif</span> <span class="n">load_type</span> <span class="o">==</span> <span class="s2">&quot;pickle&quot;</span><span class="p">:</span>
            <span class="n">pkl_path</span> <span class="o">=</span> <span class="n">base_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;.pkl&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">pkl_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Pickle file not found: </span><span class="si">{</span><span class="n">pkl_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">pkl_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">dataset</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

            <span class="n">tqdm</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Dataset loaded from: </span><span class="si">{</span><span class="n">pkl_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">dataset</span>

        <span class="k">elif</span> <span class="n">load_type</span> <span class="o">==</span> <span class="s2">&quot;tidymut&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">cls</span><span class="o">.</span><span class="n">load_by_reference</span><span class="p">(</span><span class="n">base_path</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Unsupported load_type: </span><span class="si">{</span><span class="n">load_type</span><span class="si">}</span><span class="s2">. Use &#39;dataframe&#39; or &#39;pickle&#39;&quot;</span>
            <span class="p">)</span></div>


<div class="viewcode-block" id="MutationDataset.from_dataframe">
<a class="viewcode-back" href="../../../tidymut.core.dataset.html#tidymut.core.MutationDataset.from_dataframe">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">from_dataframe</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">,</span>
        <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
        <span class="n">reference_sequences</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BaseSequence</span><span class="p">],</span>
        <span class="n">name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">specific_mutation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Type</span><span class="p">[</span><span class="n">BaseMutation</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;MutationDataset&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a MutationDataset from a DataFrame containing mutation data.</span>

<span class="sd">        This method reconstructs a MutationDataset from a flattened DataFrame representation,</span>
<span class="sd">        typically used for loading saved mutation datasets from files. The DataFrame should</span>
<span class="sd">        contain mutation information with each row representing a single mutation within</span>
<span class="sd">        mutation sets.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        df : pd.DataFrame</span>
<span class="sd">            DataFrame containing mutation data with the following required columns:</span>
<span class="sd">            - &#39;mutation_set_id&#39;: Identifier for grouping mutations into sets</span>
<span class="sd">            - &#39;reference_id&#39;: Identifier for the reference sequence</span>
<span class="sd">            - &#39;mutation_string&#39;: String representation of the mutation</span>
<span class="sd">            - &#39;position&#39;: Position of the mutation in the sequence</span>
<span class="sd">            - &#39;mutation_type&#39;: Type of mutation (&#39;amino_acid&#39;, &#39;codon_dna&#39;, &#39;codon_rna&#39;)</span>

<span class="sd">            Optional columns include:</span>
<span class="sd">            - &#39;mutation_set_name&#39;: Name of the mutation set</span>
<span class="sd">            - &#39;label&#39;: Label associated with the mutation set</span>
<span class="sd">            - &#39;wild_amino_acid&#39;: Wild-type amino acid (for amino acid mutations)</span>
<span class="sd">            - &#39;mutant_amino_acid&#39;: Mutant amino acid (for amino acid mutations)</span>
<span class="sd">            - &#39;wild_codon&#39;: Wild-type codon (for codon mutations)</span>
<span class="sd">            - &#39;mutant_codon&#39;: Mutant codon (for codon mutations)</span>
<span class="sd">            - &#39;set_*&#39;: Columns with &#39;set_&#39; prefix for mutation set metadata</span>
<span class="sd">            - &#39;mutation_*&#39;: Columns with &#39;mutation_&#39; prefix for individual mutation metadata</span>

<span class="sd">        reference_sequences : Dict[str, BaseSequence]</span>
<span class="sd">            Dictionary mapping reference sequence IDs to their corresponding BaseSequence</span>
<span class="sd">            objects. Must contain all reference sequences referenced in the DataFrame.</span>

<span class="sd">        name : Optional[str], default=None</span>
<span class="sd">            Optional name for the created MutationDataset.</span>

<span class="sd">        specific_mutation_type : Optional[BaseMutation], default=None</span>
<span class="sd">            The type of mutations to create. If None, will infer from first mutation</span>
<span class="sd">            must be provided when the mutation type is neither &#39;amino_acid&#39; nor any &#39;codon_*&#39; type.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        MutationDataset</span>
<span class="sd">            A new MutationDataset instance populated with the mutation sets and</span>
<span class="sd">            reference sequences from the DataFrame.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If the DataFrame is empty, missing required columns, or references</span>
<span class="sd">            sequences not provided in reference_sequences dict.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        - Mutations are grouped by &#39;mutation_set_id&#39; to reconstruct mutation sets</span>
<span class="sd">        - The method automatically determines the appropriate mutation set type</span>
<span class="sd">        (AminoAcidMutationSet, CodonMutationSet, or generic MutationSet) based</span>
<span class="sd">        on the mutation types within each set</span>
<span class="sd">        - Metadata is extracted from columns with &#39;set_&#39; and &#39;mutation_&#39; prefixes</span>
<span class="sd">        - Only reference sequences that are actually used in the DataFrame are</span>
<span class="sd">        added to the dataset</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; import pandas as pd</span>
<span class="sd">        &gt;&gt;&gt; from sequences import ProteinSequence</span>
<span class="sd">        &gt;&gt;&gt;</span>
<span class="sd">        &gt;&gt;&gt; # Create sample DataFrame</span>
<span class="sd">        &gt;&gt;&gt; df = pd.DataFrame({</span>
<span class="sd">        ...     &#39;mutation_set_id&#39;: [&#39;set1&#39;, &#39;set1&#39;, &#39;set2&#39;],</span>
<span class="sd">        ...     &#39;reference_id&#39;: [&#39;prot1&#39;, &#39;prot1&#39;, &#39;prot2&#39;],</span>
<span class="sd">        ...     &#39;mutation_string&#39;: [&#39;A1V&#39;, &#39;L2P&#39;, &#39;G5R&#39;],</span>
<span class="sd">        ...     &#39;position&#39;: [1, 2, 5],</span>
<span class="sd">        ...     &#39;mutation_type&#39;: [&#39;amino_acid&#39;, &#39;amino_acid&#39;, &#39;amino_acid&#39;],</span>
<span class="sd">        ...     &#39;wild_amino_acid&#39;: [&#39;A&#39;, &#39;L&#39;, &#39;G&#39;],</span>
<span class="sd">        ...     &#39;mutant_amino_acid&#39;: [&#39;V&#39;, &#39;P&#39;, &#39;R&#39;],</span>
<span class="sd">        ...     &#39;mutation_set_name&#39;: [&#39;variant1&#39;, &#39;variant1&#39;, &#39;variant2&#39;],</span>
<span class="sd">        ...     &#39;label&#39;: [&#39;pathogenic&#39;, &#39;pathogenic&#39;, &#39;benign&#39;]</span>
<span class="sd">        ... })</span>
<span class="sd">        &gt;&gt;&gt;</span>
<span class="sd">        &gt;&gt;&gt; # Define reference sequences</span>
<span class="sd">        &gt;&gt;&gt; ref_seqs = {</span>
<span class="sd">        ...     &#39;prot1&#39;: ProteinSequence(&#39;ALDEFG&#39;, name=&#39;protein1&#39;),</span>
<span class="sd">        ...     &#39;prot2&#39;: ProteinSequence(&#39;MKGLRK&#39;, name=&#39;protein2&#39;)</span>
<span class="sd">        ... }</span>
<span class="sd">        &gt;&gt;&gt;</span>
<span class="sd">        &gt;&gt;&gt; # Create MutationDataset</span>
<span class="sd">        &gt;&gt;&gt; dataset = MutationDataset.from_dataframe(df, ref_seqs, name=&quot;my_dataset&quot;)</span>
<span class="sd">        &gt;&gt;&gt; print(len(dataset.mutation_sets))</span>
<span class="sd">        2</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;DataFrame cannot be empty&quot;</span><span class="p">)</span>

        <span class="c1"># Validate required columns</span>
        <span class="n">required_cols</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;mutation_set_id&quot;</span><span class="p">,</span>
            <span class="s2">&quot;reference_id&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mutation_string&quot;</span><span class="p">,</span>
            <span class="s2">&quot;position&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mutation_type&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">missing_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">required_cols</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">missing_cols</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DataFrame missing required columns: </span><span class="si">{</span><span class="n">missing_cols</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Validate that all referenced sequences are provided</span>
        <span class="n">df_ref_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;reference_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
        <span class="n">provided_ref_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">reference_sequences</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">missing_refs</span> <span class="o">=</span> <span class="n">df_ref_ids</span> <span class="o">-</span> <span class="n">provided_ref_ids</span>
        <span class="k">if</span> <span class="n">missing_refs</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing reference sequences for IDs: </span><span class="si">{</span><span class="n">missing_refs</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Create new dataset</span>
        <span class="n">dataset</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">)</span>

        <span class="c1"># Add reference sequences</span>
        <span class="k">for</span> <span class="n">ref_id</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
            <span class="n">reference_sequences</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Adding reference sequences&quot;</span>
        <span class="p">):</span>
            <span class="k">if</span> <span class="n">ref_id</span> <span class="ow">in</span> <span class="n">df_ref_ids</span><span class="p">:</span>  <span class="c1"># Only add sequences that are actually used</span>
                <span class="n">dataset</span><span class="o">.</span><span class="n">add_reference_sequence</span><span class="p">(</span><span class="n">ref_id</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>

        <span class="c1"># Rocognize metadata columns</span>
        <span class="n">set_metadata_cols</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">col</span>
            <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span>
            <span class="k">if</span> <span class="n">col</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;set_&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">col</span> <span class="o">!=</span> <span class="s2">&quot;set_metadata&quot;</span>
        <span class="p">]</span>
        <span class="n">mutation_metadata_cols</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">col</span>
            <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span>
            <span class="k">if</span> <span class="n">col</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;mutation_&quot;</span><span class="p">)</span>
            <span class="ow">and</span> <span class="n">col</span>
            <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span>
                <span class="s2">&quot;mutation_id&quot;</span><span class="p">,</span>
                <span class="s2">&quot;mutation_type&quot;</span><span class="p">,</span>
                <span class="s2">&quot;mutation_string&quot;</span><span class="p">,</span>
                <span class="s2">&quot;mutation_category&quot;</span><span class="p">,</span>
            <span class="p">}</span>
        <span class="p">]</span>

        <span class="c1"># Group by mutation set to rebuild mutation sets</span>
        <span class="n">grouped</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;mutation_set_id&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">grouped</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Reconstructing mutation sets&quot;</span><span class="p">):</span>
            <span class="c1"># Get mutation set info</span>
            <span class="n">set_info</span> <span class="o">=</span> <span class="n">group</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">set_name</span> <span class="o">=</span> <span class="n">set_info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mutation_set_name&quot;</span><span class="p">)</span>
            <span class="n">reference_id</span> <span class="o">=</span> <span class="n">set_info</span><span class="p">[</span><span class="s2">&quot;reference_id&quot;</span><span class="p">]</span>
            <span class="n">label</span> <span class="o">=</span> <span class="n">set_info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;label&quot;</span><span class="p">)</span>

            <span class="c1"># Extract set metadata</span>
            <span class="n">set_metadata</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">col</span><span class="p">[</span><span class="mi">4</span><span class="p">:]:</span> <span class="n">value</span>
                <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">set_metadata_cols</span>
                <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notna</span><span class="p">(</span><span class="n">value</span> <span class="o">:=</span> <span class="n">set_info</span><span class="p">[</span><span class="n">col</span><span class="p">])</span>
            <span class="p">}</span>

            <span class="c1"># Create mutations from group</span>
            <span class="n">mutations</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">group</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">values</span> <span class="ow">in</span> <span class="n">group</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
                <span class="n">row_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">columns</span><span class="p">,</span> <span class="n">values</span><span class="p">))</span>
                <span class="n">mutation</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_create_mutation_from_dict</span><span class="p">(</span>
                    <span class="n">row_dict</span><span class="p">,</span> <span class="n">mutation_metadata_cols</span><span class="p">,</span> <span class="n">specific_mutation_type</span>
                <span class="p">)</span>
                <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span>

            <span class="c1"># Create appropriate mutation set type</span>
            <span class="k">if</span> <span class="n">mutations</span><span class="p">:</span>
                <span class="n">mutation_type</span> <span class="o">=</span> <span class="nb">type</span><span class="p">(</span><span class="n">mutations</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

                <span class="k">if</span> <span class="n">mutation_type</span> <span class="o">==</span> <span class="n">AminoAcidMutation</span><span class="p">:</span>
                    <span class="n">mutation_set</span> <span class="o">=</span> <span class="n">AminoAcidMutationSet</span><span class="p">(</span>
                        <span class="n">mutations</span><span class="o">=</span><span class="n">mutations</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">set_name</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="n">set_metadata</span>  <span class="c1"># type: ignore</span>
                    <span class="p">)</span>
                <span class="k">elif</span> <span class="n">mutation_type</span> <span class="o">==</span> <span class="n">CodonMutation</span><span class="p">:</span>
                    <span class="n">mutation_set</span> <span class="o">=</span> <span class="n">CodonMutationSet</span><span class="p">(</span>
                        <span class="n">mutations</span><span class="o">=</span><span class="n">mutations</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">set_name</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="n">set_metadata</span>  <span class="c1"># type: ignore</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">mutation_set</span> <span class="o">=</span> <span class="n">MutationSet</span><span class="p">(</span>
                        <span class="n">mutations</span><span class="o">=</span><span class="n">mutations</span><span class="p">,</span>
                        <span class="n">mutation_type</span><span class="o">=</span><span class="n">mutation_type</span><span class="p">,</span>
                        <span class="n">name</span><span class="o">=</span><span class="n">set_name</span><span class="p">,</span>
                        <span class="n">metadata</span><span class="o">=</span><span class="n">set_metadata</span><span class="p">,</span>
                    <span class="p">)</span>

                <span class="c1"># Add to dataset</span>
                <span class="n">dataset</span><span class="o">.</span><span class="n">add_mutation_set</span><span class="p">(</span><span class="n">mutation_set</span><span class="p">,</span> <span class="n">reference_id</span><span class="p">,</span> <span class="n">label</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">dataset</span></div>


    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_create_mutation_from_dict</span><span class="p">(</span>
        <span class="n">row_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
        <span class="n">metadata_cols</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">specific_mutation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Type</span><span class="p">[</span><span class="n">BaseMutation</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">BaseMutation</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a mutation object from a DataFrame row dict</span>
<span class="sd">        Used by the from_dataframe() method</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        row_dict: Dict[str, Any]</span>
<span class="sd">            A pandas row dict containing mutation data with the following required fields:</span>
<span class="sd">            - &#39;mutation_type&#39;: Type of mutation (&#39;amino_acid&#39;, &#39;codon_*&#39;, or other)</span>

<span class="sd">            For amino acid mutations, also requires:</span>
<span class="sd">            - &#39;wild_amino_acid&#39;: Wild-type amino acid</span>
<span class="sd">            - &#39;mutant_amino_acid&#39;: Mutant amino acid</span>

<span class="sd">            For codon mutations, also requires:</span>
<span class="sd">            - &#39;wild_codon&#39;: Wild-type codon</span>
<span class="sd">            - &#39;mutant_codon&#39;: Mutant codon</span>

<span class="sd">            For other mutations:</span>
<span class="sd">            - &#39;mutation_string&#39;: String representation of the mutation</span>

<span class="sd">            Optional fields:</span>
<span class="sd">            - Any column starting with &#39;mutation_&#39; (excluding specific reserved names)</span>
<span class="sd">            will be added as metadata to the mutation object</span>

<span class="sd">        mutation_metadata_cols : List[str]</span>
<span class="sd">            Metadata columns to be extracted and added to the mutation object</span>

<span class="sd">        specific_mutation_type : Optional[Type[BaseMutation]], default=None</span>
<span class="sd">            Only used when the mutation type is neither &#39;amino_acid&#39; nor any &#39;codon_*&#39; type.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        BaseMutation</span>
<span class="sd">            An instance of the appropriate mutation subclass:</span>
<span class="sd">            - AminoAcidMutation for &#39;amino_acid&#39; type</span>
<span class="sd">            - CodonMutation for types starting with &#39;codon_&#39;</span>
<span class="sd">            - Inferred mutation type for other types (parsed from mutation_string)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mutation_type</span> <span class="o">=</span> <span class="n">row_dict</span><span class="p">[</span><span class="s2">&quot;mutation_type&quot;</span><span class="p">]</span>
        <span class="n">position</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row_dict</span><span class="p">[</span><span class="s2">&quot;position&quot;</span><span class="p">])</span>

        <span class="c1"># Filter metadata</span>
        <span class="n">mutation_metadata</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">key</span><span class="p">[</span><span class="mi">9</span><span class="p">:]:</span> <span class="n">row_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>  <span class="c1"># Remove &quot;mutation_&quot; prefix</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">metadata_cols</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">row_dict</span><span class="p">[</span><span class="n">key</span><span class="p">])</span>
        <span class="p">}</span>

        <span class="k">if</span> <span class="n">mutation_type</span> <span class="o">==</span> <span class="s2">&quot;amino_acid&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">AminoAcidMutation</span><span class="p">(</span>
                <span class="n">wild_type</span><span class="o">=</span><span class="n">row_dict</span><span class="p">[</span><span class="s2">&quot;wild_amino_acid&quot;</span><span class="p">],</span>
                <span class="n">position</span><span class="o">=</span><span class="n">position</span><span class="p">,</span>
                <span class="n">mutant_type</span><span class="o">=</span><span class="n">row_dict</span><span class="p">[</span><span class="s2">&quot;mutant_amino_acid&quot;</span><span class="p">],</span>
                <span class="n">metadata</span><span class="o">=</span><span class="n">mutation_metadata</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">elif</span> <span class="n">mutation_type</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;codon_&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">CodonMutation</span><span class="p">(</span>
                <span class="n">wild_type</span><span class="o">=</span><span class="n">row_dict</span><span class="p">[</span><span class="s2">&quot;wild_codon&quot;</span><span class="p">],</span>
                <span class="n">position</span><span class="o">=</span><span class="n">position</span><span class="p">,</span>
                <span class="n">mutant_type</span><span class="o">=</span><span class="n">row_dict</span><span class="p">[</span><span class="s2">&quot;mutant_codon&quot;</span><span class="p">],</span>
                <span class="n">metadata</span><span class="o">=</span><span class="n">mutation_metadata</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># FIXME: need to handle other mutation types</span>
            <span class="c1"># Try to parse from mutation string as fallback</span>
            <span class="k">if</span> <span class="n">specific_mutation_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Unsupported mutation type: </span><span class="si">{</span><span class="n">mutation_type</span><span class="si">}</span><span class="s2">, &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;you must provide a specific mutation type&quot;</span>
                <span class="p">)</span>
            <span class="n">mutation_string</span> <span class="o">=</span> <span class="n">row_dict</span><span class="p">[</span><span class="s2">&quot;mutation_string&quot;</span><span class="p">]</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">MutationSet</span><span class="o">.</span><span class="n">_create_mutation</span><span class="p">(</span>
                    <span class="n">mutation_string</span><span class="p">,</span>
                    <span class="n">mutation_type</span><span class="o">=</span><span class="n">specific_mutation_type</span><span class="p">,</span>
                    <span class="n">is_zero_based</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Cannot create mutation from row: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Yuxiang Tang.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>