<!DOCTYPE html>

<html lang="en" data-content_root="../../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>tidymut.core.sequence &#8212; tidymut 0.1.0-dev documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="../../../_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="../../../_static/alabaster.css?v=27fed22d" />
    <script src="../../../_static/documentation_options.js?v=ee2d09ae"></script>
    <script src="../../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for tidymut.core.sequence</h1><div class="highlight"><pre>
<span></span><span class="c1"># tidymut/core/sequence.py</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">__future__</span><span class="w"> </span><span class="kn">import</span> <span class="n">annotations</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">abc</span><span class="w"> </span><span class="kn">import</span> <span class="n">ABC</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">TYPE_CHECKING</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNAAlphabet</span><span class="p">,</span> <span class="n">RNAAlphabet</span><span class="p">,</span> <span class="n">ProteinAlphabet</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">.codon</span><span class="w"> </span><span class="kn">import</span> <span class="n">CodonTable</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">.constants</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA_BASE_COMPLEMENTS</span><span class="p">,</span> <span class="n">RNA_BASE_COMPLEMENTS</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">.mutation</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">BaseMutation</span><span class="p">,</span>
    <span class="n">CodonMutation</span><span class="p">,</span>
    <span class="n">AminoAcidMutation</span><span class="p">,</span>
    <span class="n">MutationSet</span><span class="p">,</span>
    <span class="n">CodonMutationSet</span><span class="p">,</span>
    <span class="n">AminoAcidMutationSet</span><span class="p">,</span>
<span class="p">)</span>

<span class="k">if</span> <span class="n">TYPE_CHECKING</span><span class="p">:</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Literal</span><span class="p">,</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Type</span>

    <span class="kn">from</span><span class="w"> </span><span class="nn">.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">BaseAlphabet</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">.types</span><span class="w"> </span><span class="kn">import</span> <span class="n">SequenceType</span>


<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;BaseSequence&quot;</span><span class="p">,</span> <span class="s2">&quot;DNASequence&quot;</span><span class="p">,</span> <span class="s2">&quot;ProteinSequence&quot;</span><span class="p">,</span> <span class="s2">&quot;RNASequence&quot;</span><span class="p">]</span>


<span class="k">def</span><span class="w"> </span><span class="fm">__dir__</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
    <span class="k">return</span> <span class="n">__all__</span>


<div class="viewcode-block" id="BaseSequence">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.BaseSequence">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">BaseSequence</span><span class="p">(</span><span class="n">ABC</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Base class for biological sequences&quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">alphabet</span><span class="p">:</span> <span class="n">BaseAlphabet</span><span class="p">,</span>
        <span class="n">name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">metadata</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span> <span class="o">=</span> <span class="n">alphabet</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">validate_sequence</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata</span> <span class="ow">or</span> <span class="p">{}</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">SequenceType</span><span class="p">,</span> <span class="n">key</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SequenceType</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="nb">slice</span><span class="p">):</span>
            <span class="k">return</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">key</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="p">,</span>
                <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">key</span><span class="o">.</span><span class="n">start</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">key</span><span class="o">.</span><span class="n">stop</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">key</span><span class="o">.</span><span class="n">stop</span><span class="w"> </span><span class="ow">is</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="kc">None</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
                    <span class="k">else</span> <span class="kc">None</span>
                <span class="p">),</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">return</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">key</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="p">,</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">key</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid argument type: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">==</span> <span class="n">other</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">BaseSequence</span><span class="p">):</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">sequence</span> <span class="ow">and</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="p">)</span> <span class="o">==</span> <span class="nb">type</span><span class="p">(</span>
            <span class="n">other</span><span class="o">.</span><span class="n">alphabet</span>
        <span class="p">)</span>

<div class="viewcode-block" id="BaseSequence.get_subsequence">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.BaseSequence.get_subsequence">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_subsequence</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">SequenceType</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">end</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SequenceType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;get subsequence (0-indexed, inclusive)&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span><span class="s2">&quot;Start position must be greater than or equal to 0&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">end</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">end</span> <span class="o">&lt;</span> <span class="n">start</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;End position must be greater than or equal to start.&quot;</span><span class="p">)</span>

        <span class="n">subseq</span> <span class="o">=</span> <span class="bp">self</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">subseq</span></div>


<div class="viewcode-block" id="BaseSequence.apply_mutation">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.BaseSequence.apply_mutation">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">apply_mutation</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">SequenceType</span><span class="p">,</span>
        <span class="n">mutation</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">BaseMutation</span><span class="p">,</span> <span class="n">MutationSet</span><span class="p">,</span> <span class="n">MutationSet</span><span class="p">[</span><span class="n">BaseMutation</span><span class="p">]],</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SequenceType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Apply a mutation or set of mutations to the sequence and return a new sequence.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        mutation : Union[BaseMutation, CodonMutationSet, AminoAcidMutationSet]</span>
<span class="sd">            A single mutation or a set of mutations to apply</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        SequenceType</span>
<span class="sd">            A new sequence with the mutation(s) applied</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If mutation position is invalid or mutation is incompatible</span>
<span class="sd">        TypeError</span>
<span class="sd">            If mutation type is not supported</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Handle mutation sets (multiple mutations)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="p">(</span><span class="n">AminoAcidMutationSet</span><span class="p">,</span> <span class="n">CodonMutationSet</span><span class="p">)):</span>
            <span class="c1"># Apply mutations in reverse order of position to avoid index shifting</span>
            <span class="n">mutations</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="n">mutation</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">m</span><span class="p">:</span> <span class="n">m</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span>
            <span class="p">)</span>
            <span class="n">result_sequence</span> <span class="o">=</span> <span class="bp">self</span>

            <span class="k">for</span> <span class="n">single_mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="p">:</span>
                <span class="n">result_sequence</span> <span class="o">=</span> <span class="n">result_sequence</span><span class="o">.</span><span class="n">apply_mutation</span><span class="p">(</span><span class="n">single_mutation</span><span class="p">)</span>

            <span class="k">return</span> <span class="n">result_sequence</span>

        <span class="c1"># Handle single mutations</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">BaseMutation</span><span class="p">):</span>
            <span class="c1"># Validate mutation position and get mutation details</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">CodonMutation</span><span class="p">):</span>
                <span class="c1"># For codon mutations, check 3 bases starting at position</span>
                <span class="k">if</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">+</span> <span class="mi">2</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">):</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Codon mutation at position </span><span class="si">{</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="si">}</span><span class="s2"> extends beyond sequence length </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>

                <span class="c1"># Check mutation subtypes (DNA or RNA)</span>
                <span class="n">VALID_COMBINATIONS</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="s2">&quot;DNA&quot;</span><span class="p">:</span> <span class="n">DNAAlphabet</span><span class="p">,</span>
                    <span class="s2">&quot;Both&quot;</span><span class="p">:</span> <span class="n">DNAAlphabet</span><span class="p">,</span>
                    <span class="s2">&quot;RNA&quot;</span><span class="p">:</span> <span class="n">RNAAlphabet</span><span class="p">,</span>
                <span class="p">}</span>
                <span class="n">expected_alphabet</span> <span class="o">=</span> <span class="n">VALID_COMBINATIONS</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">seq_type</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">expected_alphabet</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="p">,</span> <span class="n">expected_alphabet</span>
                <span class="p">):</span>
                    <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Unmatching mutation subtype: </span><span class="si">{</span><span class="n">mutation</span><span class="o">.</span><span class="n">seq_type</span><span class="si">}</span><span class="s2"> with </span><span class="si">{</span><span class="n">mutation</span><span class="o">.</span><span class="n">seq_type</span><span class="si">}</span><span class="s2"> sequence&quot;</span>
                    <span class="p">)</span>

                <span class="c1"># Validate original codon matches expected</span>
                <span class="n">actual_codon</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">actual_codon</span> <span class="o">!=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">wild_codon</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Expected codon &#39;</span><span class="si">{</span><span class="n">mutation</span><span class="o">.</span><span class="n">wild_codon</span><span class="si">}</span><span class="s2">&#39; at position </span><span class="si">{</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="si">}</span><span class="s2">, &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;but found &#39;</span><span class="si">{</span><span class="n">actual_codon</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
                    <span class="p">)</span>

                <span class="c1"># Apply codon mutation (replace 3 bases)</span>
                <span class="n">new_sequence</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="p">]</span>
                    <span class="o">+</span> <span class="n">mutation</span><span class="o">.</span><span class="n">mutant_codon</span>
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">+</span> <span class="mi">3</span> <span class="p">:]</span>
                <span class="p">)</span>

            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">AminoAcidMutation</span><span class="p">):</span>
                <span class="c1"># For amino acid mutations, check single position</span>
                <span class="k">if</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">):</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Amino acid mutation position </span><span class="si">{</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="si">}</span><span class="s2"> is out of bounds for sequence of length </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>

                <span class="c1"># Validate original amino acid matches expected</span>
                <span class="n">actual_aa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">actual_aa</span> <span class="o">!=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">wild_amino_acid</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Expected amino acid &#39;</span><span class="si">{</span><span class="n">mutation</span><span class="o">.</span><span class="n">wild_amino_acid</span><span class="si">}</span><span class="s2">&#39; at position </span><span class="si">{</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="si">}</span><span class="s2">, &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;but found &#39;</span><span class="si">{</span><span class="n">actual_aa</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
                    <span class="p">)</span>

                <span class="c1"># Apply amino acid mutation (replace single position)</span>
                <span class="n">new_sequence</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="p">]</span>
                    <span class="o">+</span> <span class="n">mutation</span><span class="o">.</span><span class="n">mutant_amino_acid</span>
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">+</span> <span class="mi">1</span> <span class="p">:]</span>
                <span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Handle other BaseMutation subclasses generically</span>
                <span class="k">if</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">):</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Mutation position </span><span class="si">{</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="si">}</span><span class="s2"> is out of bounds for sequence of length </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>

                <span class="c1"># For generic mutations, we can&#39;t validate original or determine replacement length</span>
                <span class="c1"># This is a fallback for custom mutation types</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Unsupported mutation subtype: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">. &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;Only CodonMutation and AminoAcidMutation are supported.&quot;</span>
                <span class="p">)</span>

            <span class="c1"># Update metadata to track mutation</span>
            <span class="n">new_metadata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="k">if</span> <span class="s2">&quot;mutations_applied&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">new_metadata</span><span class="p">:</span>
                <span class="n">new_metadata</span><span class="p">[</span><span class="s2">&quot;mutations_applied&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="n">mutation_record</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="nb">type</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span>
                <span class="s2">&quot;mutation_type&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">type</span><span class="p">,</span>
                <span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="p">,</span>
            <span class="p">}</span>

            <span class="c1"># Add type-specific information</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">CodonMutation</span><span class="p">):</span>
                <span class="n">mutation_record</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;wild_codon&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">wild_codon</span><span class="p">,</span>
                        <span class="s2">&quot;mutant_codon&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">mutant_codon</span><span class="p">,</span>
                        <span class="s2">&quot;seq_type&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">seq_type</span><span class="p">,</span>
                    <span class="p">}</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">AminoAcidMutation</span><span class="p">):</span>
                <span class="n">mutation_record</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;wild_amino_acid&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">wild_amino_acid</span><span class="p">,</span>
                        <span class="s2">&quot;mutant_amino_acid&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">mutant_amino_acid</span><span class="p">,</span>
                        <span class="s2">&quot;effect_type&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="o">.</span><span class="n">effect_type</span><span class="p">,</span>
                    <span class="p">}</span>
                <span class="p">)</span>

            <span class="n">new_metadata</span><span class="p">[</span><span class="s2">&quot;mutations_applied&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation_record</span><span class="p">)</span>

            <span class="c1"># Create new sequence instance</span>
            <span class="k">return</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)(</span><span class="n">new_sequence</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">new_metadata</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported mutation type: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>
</div>



<div class="viewcode-block" id="ProteinSequence">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.ProteinSequence">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">ProteinSequence</span><span class="p">(</span><span class="n">BaseSequence</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Protein sequence with amino acid validation&quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">alphabet</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">ProteinAlphabet</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">metadata</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="k">if</span> <span class="n">alphabet</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">alphabet</span> <span class="o">=</span> <span class="n">ProteinAlphabet</span><span class="p">(</span><span class="n">include_stop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">alphabet</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span>

<div class="viewcode-block" id="ProteinSequence.get_residue">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.ProteinSequence.get_residue">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_residue</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">position</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get amino acid at specific position (0-indexed)&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">position</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">position</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Position </span><span class="si">{</span><span class="n">position</span><span class="si">}</span><span class="s2"> out of range (0-</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span><span class="si">}</span><span class="s2">)&quot;</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">position</span><span class="p">]</span></div>


<div class="viewcode-block" id="ProteinSequence.find_motif">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.ProteinSequence.find_motif">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">find_motif</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">motif</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find all positions where motif occurs (0-indexed)&quot;&quot;&quot;</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">motif</span> <span class="o">=</span> <span class="n">motif</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">motif</span><span class="p">,</span> <span class="n">start</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">pos</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">positions</span></div>
</div>



<div class="viewcode-block" id="RNASequence">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.RNASequence">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">RNASequence</span><span class="p">(</span><span class="n">BaseSequence</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;RNA sequence with nucleotide validation&quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">alphabet</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">RNAAlphabet</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">metadata</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="k">if</span> <span class="n">alphabet</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">alphabet</span> <span class="o">=</span> <span class="n">RNAAlphabet</span><span class="p">()</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">alphabet</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span>

<div class="viewcode-block" id="RNASequence.reverse_complement">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.RNASequence.reverse_complement">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">reverse_complement</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;RNASequence&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get reverse complement of RNA sequence&quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">rev_comp</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">RNA_BASE_COMPLEMENTS</span><span class="p">[</span><span class="n">base</span><span class="p">]</span> <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid RNA base found: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">RNASequence</span><span class="p">(</span>
            <span class="n">sequence</span><span class="o">=</span><span class="n">rev_comp</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_rc&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="RNASequence.back_transcribe">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.RNASequence.back_transcribe">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">back_transcribe</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;DNASequence&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Back-transcribe RNA sequence into DNA sequence&quot;&quot;&quot;</span>
        <span class="n">dna_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;U&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">DNASequence</span><span class="p">(</span>
            <span class="n">sequence</span><span class="o">=</span><span class="n">dna_seq</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_back_transcribe&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="RNASequence.translate">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.RNASequence.translate">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">translate</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">codon_table</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">CodonTable</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">start_at_first_met</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">stop_at_stop_codon</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">require_mod3</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">start</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">end</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ProteinSequence</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Translate RNA sequence into amino acid sequence using this codon table.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        codon_table : Optional[CodonTable], default=None</span>
<span class="sd">            Codon table to use for translation. If None, uses standard genetic code.</span>
<span class="sd">        start_at_first_met : bool, default=False</span>
<span class="sd">            Start translation at the first start codon if found.</span>
<span class="sd">        stop_at_stop_codon : bool, default=False</span>
<span class="sd">            Stop translation when a stop codon is encountered.</span>
<span class="sd">        require_mod3 : bool, default=True</span>
<span class="sd">            Whether the sequence must be a multiple of 3 in length.</span>
<span class="sd">        start : Option[int], default=None</span>
<span class="sd">            Custom 0-based start position. Overrides `start_at_first_met`.</span>
<span class="sd">        end : Option[int], default=None</span>
<span class="sd">            Custom 0-based end position. Overrides `stop_at_stop_codon`.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        ProteinSequence</span>
<span class="sd">            Translated amino acid sequence.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">aa_seq</span> <span class="o">=</span> <span class="n">translate</span><span class="p">(</span>
            <span class="n">sequence</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span>
            <span class="n">seq_type</span><span class="o">=</span><span class="s2">&quot;RNA&quot;</span><span class="p">,</span>
            <span class="n">codon_table</span><span class="o">=</span><span class="n">codon_table</span><span class="p">,</span>
            <span class="n">start_at_first_met</span><span class="o">=</span><span class="n">start_at_first_met</span><span class="p">,</span>
            <span class="n">stop_at_stop_codon</span><span class="o">=</span><span class="n">stop_at_stop_codon</span><span class="p">,</span>
            <span class="n">require_mod3</span><span class="o">=</span><span class="n">require_mod3</span><span class="p">,</span>
            <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
            <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">ProteinSequence</span><span class="p">(</span>
            <span class="n">sequence</span><span class="o">=</span><span class="n">aa_seq</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_translation&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>
</div>



<div class="viewcode-block" id="DNASequence">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.DNASequence">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">DNASequence</span><span class="p">(</span><span class="n">BaseSequence</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;DNA sequence with nucleotide validation&quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">alphabet</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">DNAAlphabet</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">metadata</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="k">if</span> <span class="n">alphabet</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">alphabet</span> <span class="o">=</span> <span class="n">DNAAlphabet</span><span class="p">()</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">alphabet</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span>

<div class="viewcode-block" id="DNASequence.reverse_complement">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.DNASequence.reverse_complement">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">reverse_complement</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;DNASequence&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get reverse complement of DNA sequence&quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">rev_comp</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">DNA_BASE_COMPLEMENTS</span><span class="p">[</span><span class="n">base</span><span class="p">]</span> <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid DNA base found: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">DNASequence</span><span class="p">(</span>
            <span class="n">sequence</span><span class="o">=</span><span class="n">rev_comp</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_rc&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="DNASequence.translate">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.DNASequence.translate">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">translate</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">codon_table</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">CodonTable</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">start_at_first_met</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">stop_at_stop_codon</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">require_mod3</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">start</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">end</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ProteinSequence</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Translate DNA sequence into amino acid sequence using this codon table.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        codon_table : Optional[CodonTable], default=None</span>
<span class="sd">            Codon table to use for translation. If None, uses standard genetic code.</span>
<span class="sd">        start_at_first_met : bool, default=False</span>
<span class="sd">            Start translation at the first start codon if found.</span>
<span class="sd">        stop_at_stop_codon : bool, default=False</span>
<span class="sd">            Stop translation when a stop codon is encountered.</span>
<span class="sd">        require_mod3 : bool, default=True</span>
<span class="sd">            Whether the sequence must be a multiple of 3 in length.</span>
<span class="sd">        start : Option[int], default=None</span>
<span class="sd">            Custom 0-based start position. Overrides `start_at_first_met`.</span>
<span class="sd">        end : Option[int], default=None</span>
<span class="sd">            Custom 0-based end position. Overrides `stop_at_stop_codon`.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        ProteinSequence</span>
<span class="sd">            Translated amino acid sequence.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">aa_seq</span> <span class="o">=</span> <span class="n">translate</span><span class="p">(</span>
            <span class="n">sequence</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span>
            <span class="n">seq_type</span><span class="o">=</span><span class="s2">&quot;DNA&quot;</span><span class="p">,</span>
            <span class="n">codon_table</span><span class="o">=</span><span class="n">codon_table</span><span class="p">,</span>
            <span class="n">start_at_first_met</span><span class="o">=</span><span class="n">start_at_first_met</span><span class="p">,</span>
            <span class="n">stop_at_stop_codon</span><span class="o">=</span><span class="n">stop_at_stop_codon</span><span class="p">,</span>
            <span class="n">require_mod3</span><span class="o">=</span><span class="n">require_mod3</span><span class="p">,</span>
            <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
            <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">ProteinSequence</span><span class="p">(</span>
            <span class="n">sequence</span><span class="o">=</span><span class="n">aa_seq</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_translation&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="DNASequence.transcribe">
<a class="viewcode-back" href="../../../tidymut.core.html#tidymut.core.sequence.DNASequence.transcribe">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">transcribe</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;RNASequence&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Transcribe DNA sequence into RNA sequence&quot;&quot;&quot;</span>
        <span class="n">rna_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;U&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">RNASequence</span><span class="p">(</span>
            <span class="n">sequence</span><span class="o">=</span><span class="n">rna_seq</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">_transcribed&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>
</div>



<span class="k">def</span><span class="w"> </span><span class="nf">translate</span><span class="p">(</span>
    <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">seq_type</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;DNA&quot;</span><span class="p">,</span> <span class="s2">&quot;RNA&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;DNA&quot;</span><span class="p">,</span>
    <span class="n">codon_table</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">CodonTable</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">start_at_first_met</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">stop_at_stop_codon</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">require_mod3</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">start</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">end</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Translate DNA or RNA sequence into amino acid sequence using this codon table.</span>
<span class="sd">    this function should not be called directly. Instead, use the DNASequence or RNASequence classes.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    sequence : str</span>
<span class="sd">        DNA or RNA sequence to translate.</span>
<span class="sd">    seq_type : Literal[&quot;DNA&quot;, &quot;RNA&quot;], default=&quot;DNA&quot;</span>
<span class="sd">    codon_table : Optional[CodonTable], default=None</span>
<span class="sd">        Codon table to use for translation. If None, uses standard genetic code.</span>
<span class="sd">    start_at_first_met : bool, default=False</span>
<span class="sd">        Start translation at the first start codon if found.</span>
<span class="sd">    stop_at_stop_codon : bool, default=False</span>
<span class="sd">        Stop translation when a stop codon is encountered.</span>
<span class="sd">    require_mod3 : bool, default=True</span>
<span class="sd">        Whether the sequence must be a multiple of 3 in length.</span>
<span class="sd">    start : Option[int], default=None</span>
<span class="sd">        Custom 0-based start position. Overrides `start_at_first_met`.</span>
<span class="sd">    end : Option[int], default=None</span>
<span class="sd">        Custom 0-based end position. Overrides `stop_at_stop_codon`.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Translated amino acid sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

    <span class="c1"># Use standard table if none provided</span>
    <span class="k">if</span> <span class="n">codon_table</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">codon_table</span> <span class="o">=</span> <span class="n">CodonTable</span><span class="o">.</span><span class="n">get_standard_table</span><span class="p">(</span><span class="n">seq_type</span><span class="o">=</span><span class="n">seq_type</span><span class="p">)</span>

    <span class="c1"># Auto detect start if `start` not provided</span>
    <span class="k">if</span> <span class="n">start</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">start_at_first_met</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">n</span> <span class="o">-</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">):</span>
                <span class="n">codon</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">i</span> <span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">codon_table</span><span class="o">.</span><span class="n">is_start_codon</span><span class="p">(</span><span class="n">codon</span><span class="p">):</span>
                    <span class="n">start</span> <span class="o">=</span> <span class="n">i</span>
                    <span class="k">break</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="s2">&quot;&quot;</span>  <span class="c1"># No start codon found</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1"># Auto detect end if `end` not provided</span>
    <span class="k">if</span> <span class="n">end</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">stop_at_stop_codon</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">n</span> <span class="o">-</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">):</span>
                <span class="n">codon</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">i</span> <span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">codon_table</span><span class="o">.</span><span class="n">is_stop_codon</span><span class="p">(</span><span class="n">codon</span><span class="p">):</span>
                    <span class="n">end</span> <span class="o">=</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">3</span>
                    <span class="k">break</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">n</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">n</span>

    <span class="n">sub_seq</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sub_seq</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">remainder</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sub_seq</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span>
        <span class="k">if</span> <span class="n">require_mod3</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Sequence length from start=</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s2"> to end=</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s2"> is not divisible by 3 &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;(remainder = </span><span class="si">{</span><span class="n">remainder</span><span class="si">}</span><span class="s2">).&quot;</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Sequence length from start=</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s2"> to end=</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s2"> is not divisible by 3. &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;Discarding </span><span class="si">{</span><span class="n">remainder</span><span class="si">}</span><span class="s2"> trailing nucleotide(s): </span><span class="si">{</span><span class="n">sub_seq</span><span class="p">[</span><span class="o">-</span><span class="n">remainder</span><span class="p">:]</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">sub_seq</span> <span class="o">=</span> <span class="n">sub_seq</span><span class="p">[:</span> <span class="nb">len</span><span class="p">(</span><span class="n">sub_seq</span><span class="p">)</span> <span class="o">-</span> <span class="n">remainder</span><span class="p">]</span>

    <span class="c1"># Translate using this codon table</span>
    <span class="n">codons</span> <span class="o">=</span> <span class="p">[</span><span class="n">sub_seq</span><span class="p">[</span><span class="n">i</span> <span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sub_seq</span><span class="p">),</span> <span class="mi">3</span><span class="p">)]</span>
    <span class="n">aa_seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">codon_table</span><span class="o">.</span><span class="n">translate_codon</span><span class="p">(</span><span class="n">codon</span><span class="p">)</span> <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">codons</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">aa_seq</span>


<span class="k">def</span><span class="w"> </span><span class="nf">load_sequences_from_fasta</span><span class="p">(</span>
    <span class="n">fasta_path</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">],</span>
    <span class="n">sequence_class</span><span class="p">:</span> <span class="n">Type</span><span class="p">[</span><span class="n">BaseSequence</span><span class="p">],</span>
    <span class="n">alphabet</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">BaseAlphabet</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">header_func</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Callable</span><span class="p">[[</span><span class="nb">str</span><span class="p">],</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">allow_duplicates</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BaseSequence</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load sequences from a FASTA file into a dictionary of BaseSequence-derived objects.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fasta_path : Union[str, Path]</span>
<span class="sd">        Path to the FASTA file</span>
<span class="sd">    sequence_class : Type[BaseSequence]</span>
<span class="sd">        A subclass of BaseSequence to instantiate (e.g. DNASequence)</span>
<span class="sd">    alphabet : Optional[BaseAlphabet], default=None</span>
<span class="sd">        Optional alphabet to validate the sequence. If None, uses default for sequence class.</span>
<span class="sd">    header_func : Optional[Callable[[str], tuple[str, str]]], default=None</span>
<span class="sd">        Function to process header line and extract (sequence_id, description).</span>
<span class="sd">        If None, uses default parsing (first word as ID, rest as description).</span>
<span class="sd">    allow_duplicates : bool, default=False</span>
<span class="sd">        If False, raises error on duplicate sequence IDs. If True, overwrites.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Dict[str, BaseSequence]</span>
<span class="sd">        Dictionary of {sequence_id: sequence_object}</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    FileNotFoundError</span>
<span class="sd">        If FASTA file doesn&#39;t exist</span>
<span class="sd">    ValueError</span>
<span class="sd">        If duplicate sequence IDs found and allow_duplicates=False</span>
<span class="sd">    TypeError</span>
<span class="sd">        If sequence_class is not a subclass of BaseSequence</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Validate inputs</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">issubclass</span><span class="p">(</span><span class="n">sequence_class</span><span class="p">,</span> <span class="n">BaseSequence</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;sequence_class must be a subclass of BaseSequence, got </span><span class="si">{</span><span class="n">sequence_class</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="n">fasta_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">fasta_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;FASTA file not found: </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Default header processing function</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">default_header_func</span><span class="p">(</span><span class="n">header_line</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract ID and description from header line (without &#39;&gt;&#39;)&quot;&quot;&quot;</span>
        <span class="n">parts</span> <span class="o">=</span> <span class="n">header_line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">maxsplit</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">seq_id</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">description</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
        <span class="k">return</span> <span class="n">seq_id</span><span class="p">,</span> <span class="n">description</span>

    <span class="c1"># Use provided header function or default</span>
    <span class="n">process_header</span> <span class="o">=</span> <span class="n">header_func</span> <span class="ow">or</span> <span class="n">default_header_func</span>

    <span class="c1"># Get default alphabet for the sequence class if not provided</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_default_alphabet</span><span class="p">():</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get default alphabet for the sequence class&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">sequence_class</span> <span class="o">==</span> <span class="n">DNASequence</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">DNAAlphabet</span><span class="p">()</span>
        <span class="k">elif</span> <span class="n">sequence_class</span> <span class="o">==</span> <span class="n">RNASequence</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">RNAAlphabet</span><span class="p">()</span>
        <span class="k">elif</span> <span class="n">sequence_class</span> <span class="o">==</span> <span class="n">ProteinSequence</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">ProteinAlphabet</span><span class="p">(</span><span class="n">include_stop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># TODO: For custom sequence classes, try to instantiate with None</span>
            <span class="c1"># and let the class handle default alphabet</span>
            <span class="c1"># but for easy, raise an error</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;sequence_class </span><span class="si">{</span><span class="n">sequence_class</span><span class="si">}</span><span class="s2"> does not have a default alphabet, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;please provide it with the alphabet parameter.&quot;</span>
            <span class="p">)</span>

    <span class="n">sequences</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BaseSequence</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">current_id</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">current_description</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">current_seq_lines</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">line_number</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">fasta_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">line_number</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

                <span class="c1"># Skip empty lines</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">):</span>
                    <span class="c1"># Save previous entry if exists</span>
                    <span class="k">if</span> <span class="n">current_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">_save_sequence</span><span class="p">(</span>
                            <span class="n">sequences</span><span class="p">,</span>
                            <span class="n">current_id</span><span class="p">,</span>
                            <span class="n">current_description</span><span class="p">,</span>
                            <span class="n">current_seq_lines</span><span class="p">,</span>
                            <span class="n">sequence_class</span><span class="p">,</span>
                            <span class="n">alphabet</span> <span class="ow">or</span> <span class="n">get_default_alphabet</span><span class="p">(),</span>
                            <span class="n">allow_duplicates</span><span class="p">,</span>
                            <span class="n">line_number</span><span class="p">,</span>
                        <span class="p">)</span>

                    <span class="c1"># Process new header</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">current_id</span><span class="p">,</span> <span class="n">current_description</span> <span class="o">=</span> <span class="n">process_header</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
                    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Error processing header at line </span><span class="si">{</span><span class="n">line_number</span><span class="si">}</span><span class="s2">: &#39;</span><span class="si">{</span><span class="n">line</span><span class="si">}</span><span class="s2">&#39;. </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">)</span>

                    <span class="n">current_seq_lines</span> <span class="o">=</span> <span class="p">[]</span>

                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Accumulate sequence lines</span>
                    <span class="c1"># Remove any whitespace and validate characters</span>
                    <span class="n">clean_line</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">())</span>  <span class="c1"># Remove all whitespace</span>
                    <span class="k">if</span> <span class="n">clean_line</span><span class="p">:</span>  <span class="c1"># Only add non-empty lines</span>
                        <span class="n">current_seq_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">clean_line</span><span class="p">)</span>

            <span class="c1"># Save last entry</span>
            <span class="k">if</span> <span class="n">current_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">_save_sequence</span><span class="p">(</span>
                    <span class="n">sequences</span><span class="p">,</span>
                    <span class="n">current_id</span><span class="p">,</span>
                    <span class="n">current_description</span><span class="p">,</span>
                    <span class="n">current_seq_lines</span><span class="p">,</span>
                    <span class="n">sequence_class</span><span class="p">,</span>
                    <span class="n">alphabet</span> <span class="ow">or</span> <span class="n">get_default_alphabet</span><span class="p">(),</span>
                    <span class="n">allow_duplicates</span><span class="p">,</span>
                    <span class="n">line_number</span><span class="p">,</span>
                <span class="p">)</span>

    <span class="k">except</span> <span class="ne">UnicodeDecodeError</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Unable to decode file </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">. Please ensure it&#39;s a valid text file.&quot;</span>
        <span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error reading FASTA file at line </span><span class="si">{</span><span class="n">line_number</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">sequences</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No valid sequences found in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">sequences</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_save_sequence</span><span class="p">(</span>
    <span class="n">sequences</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BaseSequence</span><span class="p">],</span>
    <span class="n">seq_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">description</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">seq_lines</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">sequence_class</span><span class="p">:</span> <span class="n">Type</span><span class="p">[</span><span class="n">BaseSequence</span><span class="p">],</span>
    <span class="n">alphabet</span><span class="p">:</span> <span class="n">BaseAlphabet</span><span class="p">,</span>
    <span class="n">allow_duplicates</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">line_number</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Helper function to save a sequence to the dictionary&quot;&quot;&quot;</span>
    <span class="c1"># Check for duplicate IDs</span>
    <span class="k">if</span> <span class="n">seq_id</span> <span class="ow">in</span> <span class="n">sequences</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">allow_duplicates</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Duplicate sequence ID &#39;</span><span class="si">{</span><span class="n">seq_id</span><span class="si">}</span><span class="s2">&#39; found near line </span><span class="si">{</span><span class="n">line_number</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="c1"># Join sequence lines and convert to uppercase</span>
    <span class="n">full_seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seq_lines</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>

    <span class="c1"># Skip empty sequences</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">full_seq</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Empty sequence found for ID &#39;</span><span class="si">{</span><span class="n">seq_id</span><span class="si">}</span><span class="s2">&#39;, skipping&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Create sequence object</span>
        <span class="n">sequences</span><span class="p">[</span><span class="n">seq_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">sequence_class</span><span class="p">(</span>
            <span class="n">sequence</span><span class="o">=</span><span class="n">full_seq</span><span class="p">,</span>
            <span class="n">alphabet</span><span class="o">=</span><span class="n">alphabet</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="n">seq_id</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="n">description</span><span class="p">,</span> <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="s2">&quot;fasta&quot;</span><span class="p">},</span>
        <span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error creating sequence object for ID &#39;</span><span class="si">{</span><span class="n">seq_id</span><span class="si">}</span><span class="s2">&#39;: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">tidymut</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">tidymut</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2025, Yuxiang Tang.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
    </div>

    

    
  </body>
</html>